{"cells":[{"cell_type":"markdown","id":"3c5df398","metadata":{"id":"3c5df398"},"source":["<!--NOTEBOOK_HEADER-->\n","*This notebook contains material from [PyRosetta](https://RosettaCommons.github.io/PyRosetta.notebooks);\n","content is available [on Github](https://github.com/RosettaCommons/PyRosetta.notebooks.git).*"]},{"cell_type":"markdown","id":"a647700b","metadata":{"id":"a647700b"},"source":["# Protein design\n","Keywords: pose_from_sequence(), random move, scoring move, Metropolis, assign(), Pose(), FastDesign, FastRelax, ResidueSelector, TaskFactory, TaskOperation"]},{"cell_type":"markdown","id":"e4f78667","metadata":{"id":"e4f78667"},"source":["## Overview\n","\n","In these classes we hopefully achieved the following goals:\n","\n","* Understanding Protein Structures: Before designing a protein, one must understand the current structure. PyRosetta allows for the exploration and manipulation of protein structures, providing insights into how they fold and function.\n","* Predicting Protein Folding: Proteins are made up of long chains of amino acids that fold into complex three-dimensional structures. PyRosetta can predict how changes in the sequence of amino acids will affect the way a protein folds.\n","* Designing Proteins: Once you understand the principles of protein folding and structure, you can begin to design new proteins. PyRosetta allows you to make specific changes to amino acid sequences and predict the structural and functional implications of those changes.\n","\n","We will work with the designing part by playig a bit with the helix structure of the polyA peptide from earlier. In terms of PyRosetta, we will work again with `ResidueSelectors` and `TaskOperations`.\n","\n","`ResidueSelectors` in PyRosetta are a powerful feature used to select specific residues from a protein structure based on certain criteria. These selectors provide a flexible and efficient way to identify subsets of residues that you might want to focus on for further analysis, manipulation, or scoring. Whether you are interested in a particular chain, residues with certain properties, or residues in a specific spatial region, `ResidueSelectors` offer the tools you need to make these selections and integrate them into your computational protein modeling workflows.\n","\n","`TaskOperations` in PyRosetta are components used to manipulate and configure the behavior of the Rosetta packing algorithm, which is responsible for optimizing side-chain conformations and, optionally, the backbone of a protein structure. `TaskOperation` allow you to define and customize the behavior of the packing algorithm to suit specific modeling tasks, such as protein design, mutation, or refinement.\n","\n","We can once again use FastRelax or FastDesign to do design. `FastRelax` and `FastDesign` are both protocols in PyRosetta used for optimizing protein structures, but they serve different purposes and have some key differences in their functionalities and applications.\n","\n","### FastRelax:\n","\n","- **Purpose**: `FastRelax` is primarily used for refining and relaxing protein structures. It aims to relieve any steric clashes, optimize side-chain conformations, and improve the overall energy of the protein structure.\n","- **Process**: It performs a series of minimization steps, gradually decreasing the weight of the repulsive term in the energy function to allow for larger movements initially, followed by finer adjustments.\n","- **Applications**: Commonly used for refining structures after homology modeling, loop modeling, or other structural perturbations.\n","- **Flexibility**: While it can make significant adjustments to side-chain conformations and small adjustments to backbone conformations, it does not perform extensive redesign of the protein sequence.\n","\n","### FastDesign:\n","\n","- **Purpose**: `FastDesign` extends the capabilities of `FastRelax` by incorporating the ability to redesign the protein sequence in addition to optimizing the structure.\n","- **Process**: It integrates the packing and design algorithms, allowing for simultaneous optimization of side-chain conformations and amino acid identities. Like `FastRelax`, it also uses a series of minimization steps with decreasing repulsive weight.\n","- **Applications**: Used for protein design tasks, such as designing new protein-protein interfaces, creating novel protein folds, or improving protein stability.\n","- **Flexibility**: Provides extensive capabilities for both structural optimization and sequence redesign, making it a powerful tool for protein engineering.\n","\n","### Main Differences:\n","\n","1. **Sequence Design**: The most significant difference between `FastRelax` and `FastDesign` is the ability of `FastDesign` to redesign the protein sequence. `FastRelax` focuses solely on optimizing the structure.\n","2. **Applications**: `FastRelax` is typically used for structure refinement, while `FastDesign` is used for more extensive protein design and engineering tasks.\n","3. **Flexibility**: `FastDesign` offers more flexibility and capabilities for manipulating both the structure and sequence of a protein."]},{"cell_type":"markdown","id":"faafc753","metadata":{"id":"faafc753"},"source":["## Basic $\\alpha$-helix redesign"]},{"cell_type":"code","source":["\n","!pip install pyrosettacolabsetup\n","import pyrosettacolabsetup; pyrosettacolabsetup.install_pyrosetta()\n","from pyrosetta import *\n","from pyrosetta.rosetta import *\n","from pyrosetta.teaching import *\n","import pyrosetta.toolbox\n","#pyrosetta.init()\n","\n","pyrosetta.init(\"-ex1 -ex2aro\")\n","!pip install py3Dmol\n","import py3Dmol\n"],"metadata":{"id":"1nEQMw1-Uzi6","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1727253724517,"user_tz":-120,"elapsed":122684,"user":{"displayName":"Aurelio Angel Moya Garcia","userId":"06239010930884764093"}},"outputId":"82f80f15-32f8-45a1-e41f-5717798d2946"},"id":"1nEQMw1-Uzi6","execution_count":1,"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting pyrosettacolabsetup\n","  Downloading pyrosettacolabsetup-1.0.9-py3-none-any.whl.metadata (294 bytes)\n","Downloading pyrosettacolabsetup-1.0.9-py3-none-any.whl (4.9 kB)\n","Installing collected packages: pyrosettacolabsetup\n","Successfully installed pyrosettacolabsetup-1.0.9\n","Mounted at /content/google_drive\n","\n","Note that USE OF PyRosetta FOR COMMERCIAL PURPOSES REQUIRE PURCHASE OF A LICENSE.\n","See https://github.com/RosettaCommons/rosetta/blob/main/LICENSE.md or email license@uw.edu for details.\n","\n","Looking for compatible PyRosetta wheel file at google-drive/PyRosetta/colab.bin//wheels...\n","Found compatible wheel: /content/google_drive/MyDrive/PyRosetta/colab.bin/wheels//content/google_drive/MyDrive/PyRosetta/colab.bin/wheels/pyrosetta-2024.19+release.a34b73c40f-cp310-cp310-linux_x86_64.whl\n","\n","\n","┌──────────────────────────────────────────────────────────────────────────────┐\n","│                                 PyRosetta-4                                  │\n","│              Created in JHU by Sergey Lyskov and PyRosetta Team              │\n","│              (C) Copyright Rosetta Commons Member Institutions               │\n","│                                                                              │\n","│ NOTE: USE OF PyRosetta FOR COMMERCIAL PURPOSES REQUIRE PURCHASE OF A LICENSE │\n","│         See LICENSE.PyRosetta.md or email license@uw.edu for details         │\n","└──────────────────────────────────────────────────────────────────────────────┘\n","PyRosetta-4 2024 [Rosetta PyRosetta4.MinSizeRel.python310.ubuntu 2024.19+release.a34b73c40fe9c61558d566d6a63f803cfb15a4fc 2024-05-02T16:22:03] retrieved from: http://www.pyrosetta.org\n","core.init: Checking for fconfig files in pwd and ./rosetta/flags\n","core.init: Rosetta version: PyRosetta4.MinSizeRel.python310.ubuntu r381 2024.19+release.a34b73c40f a34b73c40fe9c61558d566d6a63f803cfb15a4fc http://www.pyrosetta.org 2024-05-02T16:22:03\n","core.init: Rosetta extras: []\n","core.init: command: PyRosetta -ex1 -ex2aro -database /usr/local/lib/python3.10/dist-packages/pyrosetta/database\n","basic.random.init_random_generator: 'RNG device' seed mode, using '/dev/urandom', seed=1925841855 seed_offset=0 real_seed=1925841855\n","basic.random.init_random_generator: RandomGenerator:init: Normal mode, seed=1925841855 RG_type=mt19937\n","Collecting py3Dmol\n","  Downloading py3Dmol-2.4.0-py2.py3-none-any.whl.metadata (1.9 kB)\n","Downloading py3Dmol-2.4.0-py2.py3-none-any.whl (7.0 kB)\n","Installing collected packages: py3Dmol\n","Successfully installed py3Dmol-2.4.0\n"]}]},{"cell_type":"markdown","id":"cd6c58d6","metadata":{"id":"cd6c58d6"},"source":["Initialize pyrosetta.\n","- The `-ex1` option in PyRosetta is used to increase the accuracy of side-chain modeling by enabling additional rotamer sampling for chi1 angles of amino acid side chains during simulations.\n","- The `ex2aro`is used to enhance the sampling of aromatic side chains during protein structure prediction and design simulations. When the -ex2aro option is enabled, PyRosetta performs additional sampling of the chi2 dihedral angles for aromatic residues. This results in a more thorough exploration of the conformational space of these residues, which can lead to more accurate and reliable modeling of protein structures at the cost of additional computational resources."]},{"cell_type":"code","source":["from google.colab import drive\n","drive.mount('/content/drive')"],"metadata":{"id":"s_4DsvuCWUDT"},"id":"s_4DsvuCWUDT","execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["### Function to visualise poses"],"metadata":{"id":"lNtNzuMNcCsa"},"id":"lNtNzuMNcCsa"},{"cell_type":"code","source":["import tempfile\n","\n","\n","# Function to display the pose\n","def display_protein(pose):\n","    with tempfile.NamedTemporaryFile(suffix='.pdb') as tmp:\n","        pose.dump_pdb(tmp.name) #Indented this line\n","        with open(tmp.name, 'r') as pdb_file: #Indented this line\n","            pdb_string = pdb_file.read() #Indented this line\n","    viewer = py3Dmol.view(width=800, height=600)\n","    viewer.addModel(pdb_string, 'pdb')\n","    # viewer.setStyle({'cartoon': {'color': 'spectrum'}})\n","    viewer.setStyle({'stick': {'colorscheme': 'greenCarbon'}})\n","    viewer.zoomTo()\n","    viewer.show()\n","\n","\n","\n","# Save the pose to a temporary PDB file and read it as a string\n","#with tempfile.NamedTemporaryFile(suffix='.pdb') as tmp:\n","#    polyA.dump_pdb(tmp.name)\n"," #   with open(tmp.name, 'r') as pdb_file:\n"," #       pdb_string = pdb_file.read()\n","\n"],"metadata":{"id":"OovSvxxOcJyF"},"id":"OovSvxxOcJyF","execution_count":null,"outputs":[]},{"cell_type":"markdown","id":"d5f382dc","metadata":{"id":"d5f382dc"},"source":["### Building the pose\n","We start from the folded polyA peptide stored in a variable called \"start_pose\""]},{"cell_type":"code","execution_count":null,"id":"f0b2d129","metadata":{"id":"f0b2d129"},"outputs":[],"source":["#start_pose = pyrosetta.pose_from_sequence('A' * 10)\n","# Better to start from the folded peptide we obtained earlier\n","\n","scorefxn = pyrosetta.create_score_function(\"ref2015_cart.wts\")\n","start_pose = pose_from_pdb(\"/content/google_drive/MyDrive/BIP_24-25/5-Protein_design/mc_final.pdb\") # Call the function correctly\n","pose = start_pose.clone()\n","pose = start_pose.clone()\n"]},{"cell_type":"code","execution_count":null,"id":"33748443","metadata":{"id":"33748443"},"outputs":[],"source":["# measure energy level\n","scorefxn(pose)"]},{"cell_type":"markdown","id":"0e4733a1","metadata":{"id":"0e4733a1"},"source":["We will use the function above to inspect the initial conformation of the polyA peptide."]},{"cell_type":"code","execution_count":null,"id":"4858b78f","metadata":{"id":"4858b78f"},"outputs":[],"source":["display_protein(pose)"]},{"cell_type":"markdown","id":"832bb2d5","metadata":{"id":"832bb2d5"},"source":["### Insert mutations"]},{"cell_type":"markdown","id":"ef3fbe45","metadata":{"id":"ef3fbe45"},"source":["Define the mutations you want to introduce. We need to write them in a resfile.\n","\n","In PyRosetta, a resfile is used to specify which residues should be mutated and to what amino acid they should be changed. The resfile format allows for a lot of flexibility and control over the design process.\n","\n","Here is an example of a resfile that would mutate positions 3 and 7 of a 10 amino acid polyalanine peptide. In this example, we will mutate position 3 to aspartate (D) and position 7 to leucine (L).\n","\n","```\n","NATRO\n","start\n","\n","3 A PIKAA D\n","7 A PIKAA L\n","```\n","\n","* NATRO: This command specifies that all residues not mentioned explicitly should not be repacked or mutated (they remain \"naturally\" as they are).\n","* start: This indicates the beginning of the residue-specific commands.\n","* 3 A PIKAA D: This line tells PyRosetta to mutate residue number 3 in chain A to aspartate. PIKAA stands for \"point mutation using the indicated amino acid.\"\n","* 7 A PIKAA L: This line tells PyRosetta to mutate residue number 7 in chain A to leucine.\n","\n","Check out the design proposed in the `design.resfile`."]},{"cell_type":"code","execution_count":null,"id":"6f1cb47b","metadata":{"id":"6f1cb47b"},"outputs":[],"source":["resfile = \"/content/google_drive/MyDrive/BIP_24-25/5-Protein_design/design.resfile\""]},{"cell_type":"markdown","id":"b62ca769","metadata":{"id":"b62ca769"},"source":["Now we can setup the TaskOperations for the `FastRelax` mover. These tell `FastRelax` which residues to design or repack during the packer steps in `FastRelax`.\n"]},{"cell_type":"markdown","source":["In PyRosetta, **TaskOperations** and **FastRelax** are crucial for controlling how the protein structure is modified, especially during tasks like structure optimization and repacking.\n","\n","\n","\n","**TaskOperations** define specific rules and operations that guide how residues in a protein are manipulated during structure refinement, repacking, and design processes. They specify which residues are allowed to be mutated, which rotamers to sample, and how packing should proceed. In your code:\n","\n","- `TaskFactory`: This object collects all the **TaskOperations**. It's like a manager that bundles all the operations you want to apply to your structure.\n","  \n","- `InitializeFromCommandline()`: This operation sets up the packing task based on any command-line options provided during PyRosetta initialization. It helps initialize the task with default settings.\n","\n","- `IncludeCurrent()`: This operation ensures that the current rotamers (the conformations of side chains) of residues are included in the packing process. It ensures that the current structure is among the options when PyRosetta evaluates side-chain conformations.\n","\n","- `ReadResfile(resfile)`: This reads a **resfile**, a text file that specifies custom rules about which residues should be repacked or redesigned. You can use it to fine-tune how specific regions of the protein are handled.\n","\n","By using these **TaskOperations**, you can define which residues in your protein should be flexible, which ones should remain fixed, or even guide specific mutation and repacking strategies. The `packer_task` created from this allows you to inspect the task and see how it's being applied to your protein (`pose`).\n","\n","\n","\n","**FastRelax** is a powerful mover in PyRosetta designed for fast and efficient protein structure refinement. It combines multiple cycles of energy minimization and side-chain repacking to produce low-energy protein models. It’s a faster variant of the standard **Relax** protocol, but it still provides high-quality results.\n","\n","- **Minimization**: FastRelax performs gradient-based energy minimization, adjusting the protein's backbone and side chains to lower the overall energy.\n","  \n","- **Repacking**: During FastRelax, side chains are repacked to explore different rotamer conformations. This is where **TaskOperations** come into play: they control which residues get repacked, how they are sampled, and what restrictions (e.g., from a resfile) should be applied.\n","\n","FastRelax typically works in several cycles, alternating between repacking and minimizing to drive the protein structure to a local energy minimum.\n","\n","In the code below, the **TaskFactory** you set up defines the rules for how the residues will be repacked during the **FastRelax** protocol, ensuring that the current rotamers are considered and that any custom residue rules from the resfile are respected.\n","\n","\n","The **PackerTask** in PyRosetta defines the specific set of operations and rules used during **repacking** or **design** of side chains in a protein structure. It's essentially a blueprint that controls how the side chains (rotamers) of residues in a protein will be sampled, repacked, or redesigned.\n","\n","##### Key aspects of a **PackerTask**:\n","\n","1. **Repacking**: The PackerTask dictates which residues' side chains can be repacked, meaning their rotamer conformations are optimized while keeping the amino acid identity fixed. This is useful for refining the structure without changing its overall sequence.\n","\n","2. **Design**: In addition to repacking, a PackerTask can also be used to redesign certain residues, meaning the side chain can change to a different amino acid entirely. The rules for this can be specified by the resfile or other task operations.\n","\n","3. **TaskOperations**: A PackerTask is generated by applying **TaskOperations**, which control:\n","   - Which residues are allowed to be modified (packed, repacked, or designed).\n","   - Whether the current side chain conformation is included in the repacking process.\n","   - Any restrictions on amino acid substitutions or packing regions.\n","\n","4. **Residue-level control**: For each residue in the protein, the PackerTask can specify whether it is allowed to move, which rotamers can be sampled, and whether the side chain should be repacked, designed, or remain fixed.\n","\n","\n","In summary, the **PackerTask** is a detailed set of instructions that defines how side chains are handled during repacking or design, ensuring that the specified residues follow the rules set by the task operations and resfile during processes like FastRelax.\n"],"metadata":{"id":"WLP8vEehfhMu"},"id":"WLP8vEehfhMu"},{"cell_type":"code","execution_count":null,"id":"b0d148d1","metadata":{"id":"b0d148d1"},"outputs":[],"source":["# The task factory accepts all the task operations\n","tf = pyrosetta.rosetta.core.pack.task.TaskFactory()\n","\n","# These are pretty standard\n","tf.push_back(pyrosetta.rosetta.core.pack.task.operation.InitializeFromCommandline())\n","tf.push_back(pyrosetta.rosetta.core.pack.task.operation.IncludeCurrent())\n","\n","\n","# Include the resfile\n","tf.push_back(pyrosetta.rosetta.core.pack.task.operation.ReadResfile(resfile))\n","\n","# Convert the task factory into a PackerTask to take a look at it\n","packer_task = tf.create_task_and_apply_taskoperations(pose)\n","# View the PackerTask\n","print(packer_task)"]},{"cell_type":"markdown","id":"ae8332b2","metadata":{"id":"ae8332b2"},"source":[" The PackerTask looks as intended!\n","\n"," Now we can set up a `MoveMap` to specify which torsions are free to minimize during the minimization steps of the `FastDesign` mover.\n","\n"," A `MoveMap` in PyRosetta is a data structure that defines which degrees of freedom (backbone, side chains, or rigid-body jumps) in a protein structure can be moved or optimized during refinement or minimization processes. It's used to control which parts of the protein are flexible and which are fixed during various protocols such as **FastRelax**, **Minimization**, and **CartesianMinimization**.\n","\n","##### Components of a **MoveMap**:\n","\n","1. **Backbone flexibility (bb)**: The protein's backbone consists of the φ (phi) and ψ (psi) torsion angles for each residue. Setting the backbone to be flexible means these torsion angles can change during the optimization.\n","   \n","2. **Side-chain flexibility (chi)**: The side-chain dihedral angles (χ angles) can also be adjusted. Allowing chi flexibility means that side-chain conformations (rotamers) can be optimized during refinement.\n","\n","3. **Jump flexibility (jump)**: Proteins or complexes that contain multiple domains or chains connected by rigid-body jumps (rigid-body transformations between two parts of a structure) can have their relative orientations changed if jumps are flexible.\n","\n","### Breakdown of the code:\n","\n","```python\n","mm = pyrosetta.rosetta.core.kinematics.MoveMap()\n","mm.set_bb(True)   # Allow backbone torsions (phi, psi) to move\n","mm.set_chi(True)  # Allow side-chain torsions (chi) to move\n","mm.set_jump(True) # Allow rigid-body jumps to move\n","```\n","\n","- `mm.set_bb(True)`: This command allows the **backbone** torsions of the protein to move during optimization. If set to `False`, the backbone remains fixed.\n","  \n","- `mm.set_chi(True)`: This enables movement in the **side-chain dihedral angles**. If set to `False`, the side chains remain fixed.\n","\n","- `mm.set_jump(True)`: This allows **rigid-body jumps** (which define the relative positioning of chains or domains) to be altered.\n","\n","##### How the **MoveMap** is used:\n","\n","\n","In the code below, you are allowing **all parts of the protein**—the backbone, side chains, and jumps—to move, making the entire structure flexible during whatever optimization protocol you're running.\n"]},{"cell_type":"code","execution_count":null,"id":"51051a95","metadata":{"id":"51051a95"},"outputs":[],"source":["mm = pyrosetta.rosetta.core.kinematics.MoveMap()\n","mm.set_bb(True)\n","mm.set_chi(True)\n","mm.set_jump(True)"]},{"cell_type":"markdown","id":"2ac89f0c","metadata":{"id":"2ac89f0c"},"source":["Set up `FastDesign`"]},{"cell_type":"code","execution_count":null,"id":"43d83078","metadata":{"id":"43d83078"},"outputs":[],"source":["rel_design = pyrosetta.rosetta.protocols.relax.FastRelax(scorefxn_in=scorefxn, standard_repeats=1, script_file=\"MonomerDesign2019\")\n","rel_design.cartesian(True)\n","rel_design.set_task_factory(tf)\n","rel_design.set_movemap(mm)\n","rel_design.minimize_bond_angles(True)\n","rel_design.minimize_bond_lengths(True)"]},{"cell_type":"markdown","id":"5b604d01","metadata":{"id":"5b604d01"},"source":[" Run the `FastDesign` mover. This is fast"]},{"cell_type":"code","execution_count":null,"id":"737ed6c0","metadata":{"id":"737ed6c0"},"outputs":[],"source":["d3l7_pose = pose.clone()  # Create a copy of the original pose\n","%time rel_design.apply(d3l7_pose)"]},{"cell_type":"markdown","id":"1e52bf3d","metadata":{"id":"1e52bf3d"},"source":["We can see the result of the design."]},{"cell_type":"code","source":["display_protein(d3l7_pose)"],"metadata":{"id":"Yh-MM7Oubxxt"},"id":"Yh-MM7Oubxxt","execution_count":null,"outputs":[]},{"cell_type":"markdown","id":"733365f6","metadata":{"id":"733365f6"},"source":["Since the peptide was already folded, we don't see major changes in its conformation. However, we can compare the relative stability of the design by measuring the Rosetta energy before and after design.\n"]},{"cell_type":"code","execution_count":null,"id":"0e12686c","metadata":{"id":"0e12686c"},"outputs":[],"source":["before_design = scorefxn(pose)\n","print(before_design)\n","energy_design = scorefxn(d3l7_pose)\n","print(energy_design)"]},{"cell_type":"markdown","source":["**Exercise:** Apparently, we have stabilised the helix with our design. How is this possible?"],"metadata":{"id":"OcaMNAj9kBJR"},"id":"OcaMNAj9kBJR"},{"cell_type":"markdown","id":"4435f7c5","metadata":{"id":"4435f7c5"},"source":["To properly measure the energy level of the no-design peptide, we need to subject it to the same energy minimization.\n"]},{"cell_type":"code","execution_count":null,"id":"1ba0e1cb","metadata":{"id":"1ba0e1cb"},"outputs":[],"source":["# new design resfile\n","nodesign_resfile = \"/content/google_drive/MyDrive/BIP_24-25/5-Protein_design/nodesign.resfile\"\n","\n","#reset pose to the initial conformation\n","pose = start_pose.clone()\n","\n","\n","# The task factory accepts all the task operations\n","tf = pyrosetta.rosetta.core.pack.task.TaskFactory()\n","\n","# These are pretty standard\n","tf.push_back(pyrosetta.rosetta.core.pack.task.operation.InitializeFromCommandline())\n","tf.push_back(pyrosetta.rosetta.core.pack.task.operation.IncludeCurrent())\n","\n","\n","# Include the resfile\n","tf.push_back(pyrosetta.rosetta.core.pack.task.operation.ReadResfile(nodesign_resfile))\n","\n","# Convert the task factory into a PackerTask to take a look at it\n","packer_task = tf.create_task_and_apply_taskoperations(pose)\n","# View the PackerTask\n","print(packer_task)\n","\n","\n","mm = pyrosetta.rosetta.core.kinematics.MoveMap()\n","mm.set_bb(True)\n","mm.set_chi(True)\n","mm.set_jump(True)\n","\n","rel_design = pyrosetta.rosetta.protocols.relax.FastRelax(scorefxn_in=scorefxn, standard_repeats=1, script_file=\"MonomerDesign2019\")\n","rel_design.cartesian(True)\n","rel_design.set_task_factory(tf)\n","rel_design.set_movemap(mm)\n","rel_design.minimize_bond_angles(True)\n","rel_design.minimize_bond_lengths(True)\n","\n","%time rel_design.apply(pose)\n","\n"]},{"cell_type":"code","execution_count":null,"id":"f9746fef","metadata":{"id":"f9746fef"},"outputs":[],"source":["energy_nodesign = scorefxn(pose)\n","print(energy_nodesign)"]},{"cell_type":"code","execution_count":null,"id":"b10e9aee","metadata":{"id":"b10e9aee"},"outputs":[],"source":["print(\"deltaE =\", energy_design - energy_nodesign)"]},{"cell_type":"markdown","id":"30b31db5","metadata":{"id":"30b31db5"},"source":["**Exercise. Mess with the polyA peptide. Explore the mutations that destabilize the $\\alpha$-helix.**"]}],"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.11.5"},"colab":{"provenance":[]}},"nbformat":4,"nbformat_minor":5}