{"cells":[{"cell_type":"markdown","id":"3c5df398","metadata":{"id":"3c5df398"},"source":["<!--NOTEBOOK_HEADER-->\n","*This notebook contains material from [PyRosetta](https://RosettaCommons.github.io/PyRosetta.notebooks);\n","content is available [on Github](https://github.com/RosettaCommons/PyRosetta.notebooks.git).*"]},{"cell_type":"markdown","id":"a647700b","metadata":{"id":"a647700b"},"source":["# Protein design\n","Keywords: pose_from_sequence(), random move, scoring move, Metropolis, assign(), Pose(), FastDesign, FastRelax, ResidueSelector, TaskFactory, TaskOperation"]},{"cell_type":"markdown","id":"e4f78667","metadata":{"id":"e4f78667"},"source":["## Overview\n","\n","In these classes we hopefully achieved the following goals:\n","\n","* Understanding Protein Structures: Before designing a protein, one must understand the current structure. PyRosetta allows for the exploration and manipulation of protein structures, providing insights into how they fold and function.\n","* Predicting Protein Folding: Proteins are made up of long chains of amino acids that fold into complex three-dimensional structures. PyRosetta can predict how changes in the sequence of amino acids will affect the way a protein folds.\n","* Designing Proteins: Once you understand the principles of protein folding and structure, you can begin to design new proteins. PyRosetta allows you to make specific changes to amino acid sequences and predict the structural and functional implications of those changes.\n","\n","We will work with the designing part by playig a bit with the helix structure of the polyA peptide from earlier. In terms of PyRosetta, we will work again with `ResidueSelectors` and `TaskOperations`.\n","\n","`ResidueSelectors` in PyRosetta are a powerful feature used to select specific residues from a protein structure based on certain criteria. These selectors provide a flexible and efficient way to identify subsets of residues that you might want to focus on for further analysis, manipulation, or scoring. Whether you are interested in a particular chain, residues with certain properties, or residues in a specific spatial region, `ResidueSelectors` offer the tools you need to make these selections and integrate them into your computational protein modeling workflows.\n","\n","`TaskOperations` in PyRosetta are components used to manipulate and configure the behavior of the Rosetta packing algorithm, which is responsible for optimizing side-chain conformations and, optionally, the backbone of a protein structure. `TaskOperation` allow you to define and customize the behavior of the packing algorithm to suit specific modeling tasks, such as protein design, mutation, or refinement.\n","\n","We can once again use FastRelax or FastDesign to do design. `FastRelax` and `FastDesign` are both protocols in PyRosetta used for optimizing protein structures, but they serve different purposes and have some key differences in their functionalities and applications.\n","\n","### FastRelax:\n","\n","- **Purpose**: `FastRelax` is primarily used for refining and relaxing protein structures. It aims to relieve any steric clashes, optimize side-chain conformations, and improve the overall energy of the protein structure.\n","- **Process**: It performs a series of minimization steps, gradually decreasing the weight of the repulsive term in the energy function to allow for larger movements initially, followed by finer adjustments.\n","- **Applications**: Commonly used for refining structures after homology modeling, loop modeling, or other structural perturbations.\n","- **Flexibility**: While it can make significant adjustments to side-chain conformations and small adjustments to backbone conformations, it does not perform extensive redesign of the protein sequence.\n","\n","### FastDesign:\n","\n","- **Purpose**: `FastDesign` extends the capabilities of `FastRelax` by incorporating the ability to redesign the protein sequence in addition to optimizing the structure.\n","- **Process**: It integrates the packing and design algorithms, allowing for simultaneous optimization of side-chain conformations and amino acid identities. Like `FastRelax`, it also uses a series of minimization steps with decreasing repulsive weight.\n","- **Applications**: Used for protein design tasks, such as designing new protein-protein interfaces, creating novel protein folds, or improving protein stability.\n","- **Flexibility**: Provides extensive capabilities for both structural optimization and sequence redesign, making it a powerful tool for protein engineering.\n","\n","### Main Differences:\n","\n","1. **Sequence Design**: The most significant difference between `FastRelax` and `FastDesign` is the ability of `FastDesign` to redesign the protein sequence. `FastRelax` focuses solely on optimizing the structure.\n","2. **Applications**: `FastRelax` is typically used for structure refinement, while `FastDesign` is used for more extensive protein design and engineering tasks.\n","3. **Flexibility**: `FastDesign` offers more flexibility and capabilities for manipulating both the structure and sequence of a protein."]},{"cell_type":"markdown","id":"faafc753","metadata":{"id":"faafc753"},"source":["## Basic $\\alpha$-helix redesign"]},{"cell_type":"code","source":["\n","!pip install pyrosettacolabsetup\n","import pyrosettacolabsetup; pyrosettacolabsetup.install_pyrosetta()\n","from pyrosetta import *\n","from pyrosetta.rosetta import *\n","from pyrosetta.teaching import *\n","import pyrosetta.toolbox\n","#pyrosetta.init()\n","\n","pyrosetta.init(\"-ex1 -ex2aro\")\n","!pip install py3Dmol\n","import py3Dmol\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"1nEQMw1-Uzi6","outputId":"e809feed-ddc2-4b8f-9103-ca66eab5ca37","executionInfo":{"status":"ok","timestamp":1727253641699,"user_tz":-120,"elapsed":120186,"user":{"displayName":"Aurelio Angel Moya Garcia","userId":"06239010930884764093"}}},"id":"1nEQMw1-Uzi6","execution_count":1,"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting pyrosettacolabsetup\n","  Downloading pyrosettacolabsetup-1.0.9-py3-none-any.whl.metadata (294 bytes)\n","Downloading pyrosettacolabsetup-1.0.9-py3-none-any.whl (4.9 kB)\n","Installing collected packages: pyrosettacolabsetup\n","Successfully installed pyrosettacolabsetup-1.0.9\n","Mounted at /content/google_drive\n","\n","Note that USE OF PyRosetta FOR COMMERCIAL PURPOSES REQUIRE PURCHASE OF A LICENSE.\n","See https://github.com/RosettaCommons/rosetta/blob/main/LICENSE.md or email license@uw.edu for details.\n","\n","Looking for compatible PyRosetta wheel file at google-drive/PyRosetta/colab.bin//wheels...\n","Found compatible wheel: /content/google_drive/MyDrive/PyRosetta/colab.bin/wheels//content/google_drive/MyDrive/PyRosetta/colab.bin/wheels/pyrosetta-2024.19+release.a34b73c40f-cp310-cp310-linux_x86_64.whl\n","\n","\n","┌──────────────────────────────────────────────────────────────────────────────┐\n","│                                 PyRosetta-4                                  │\n","│              Created in JHU by Sergey Lyskov and PyRosetta Team              │\n","│              (C) Copyright Rosetta Commons Member Institutions               │\n","│                                                                              │\n","│ NOTE: USE OF PyRosetta FOR COMMERCIAL PURPOSES REQUIRE PURCHASE OF A LICENSE │\n","│         See LICENSE.PyRosetta.md or email license@uw.edu for details         │\n","└──────────────────────────────────────────────────────────────────────────────┘\n","PyRosetta-4 2024 [Rosetta PyRosetta4.MinSizeRel.python310.ubuntu 2024.19+release.a34b73c40fe9c61558d566d6a63f803cfb15a4fc 2024-05-02T16:22:03] retrieved from: http://www.pyrosetta.org\n","core.init: Checking for fconfig files in pwd and ./rosetta/flags\n","core.init: Rosetta version: PyRosetta4.MinSizeRel.python310.ubuntu r381 2024.19+release.a34b73c40f a34b73c40fe9c61558d566d6a63f803cfb15a4fc http://www.pyrosetta.org 2024-05-02T16:22:03\n","core.init: Rosetta extras: []\n","core.init: command: PyRosetta -ex1 -ex2aro -database /usr/local/lib/python3.10/dist-packages/pyrosetta/database\n","basic.random.init_random_generator: 'RNG device' seed mode, using '/dev/urandom', seed=-822719513 seed_offset=0 real_seed=-822719513\n","basic.random.init_random_generator: RandomGenerator:init: Normal mode, seed=-822719513 RG_type=mt19937\n","Collecting py3Dmol\n","  Downloading py3Dmol-2.4.0-py2.py3-none-any.whl.metadata (1.9 kB)\n","Downloading py3Dmol-2.4.0-py2.py3-none-any.whl (7.0 kB)\n","Installing collected packages: py3Dmol\n","Successfully installed py3Dmol-2.4.0\n"]}]},{"cell_type":"code","execution_count":null,"id":"44fdde87","metadata":{"id":"44fdde87"},"outputs":[],"source":["import os\n","import logging\n","logging.basicConfig(level=logging.INFO)\n","from pyrosetta import *\n","from pyrosetta.rosetta import *\n","from pyrosetta.teaching import *\n","import pyrosetta.toolbox\n","pyrosetta.init(\"-ex1 -ex2aro\")"]},{"cell_type":"markdown","id":"cd6c58d6","metadata":{"id":"cd6c58d6"},"source":["Initialize pyrosetta.\n","- The `-ex1` option in PyRosetta is used to increase the accuracy of side-chain modeling by enabling additional rotamer sampling for chi1 angles of amino acid side chains during simulations.\n","- The `ex2aro`is used to enhance the sampling of aromatic side chains during protein structure prediction and design simulations. When the -ex2aro option is enabled, PyRosetta performs additional sampling of the chi2 dihedral angles for aromatic residues. This results in a more thorough exploration of the conformational space of these residues, which can lead to more accurate and reliable modeling of protein structures at the cost of additional computational resources."]},{"cell_type":"code","source":["from google.colab import drive\n","drive.mount('/content/drive')"],"metadata":{"id":"s_4DsvuCWUDT"},"id":"s_4DsvuCWUDT","execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["### Function to visualise poses"],"metadata":{"id":"lNtNzuMNcCsa"},"id":"lNtNzuMNcCsa"},{"cell_type":"code","source":["import tempfile\n","\n","\n","# Function to display the pose\n","def display_protein(pose):\n","    with tempfile.NamedTemporaryFile(suffix='.pdb') as tmp:\n","        pose.dump_pdb(tmp.name) #Indented this line\n","        with open(tmp.name, 'r') as pdb_file: #Indented this line\n","            pdb_string = pdb_file.read() #Indented this line\n","    viewer = py3Dmol.view(width=800, height=600)\n","    viewer.addModel(pdb_string, 'pdb')\n","    # viewer.setStyle({'cartoon': {'color': 'spectrum'}})\n","    viewer.setStyle({'stick': {'colorscheme': 'greenCarbon'}})\n","    viewer.zoomTo()\n","    viewer.show()\n","\n","\n","\n","# Save the pose to a temporary PDB file and read it as a string\n","#with tempfile.NamedTemporaryFile(suffix='.pdb') as tmp:\n","#    polyA.dump_pdb(tmp.name)\n"," #   with open(tmp.name, 'r') as pdb_file:\n"," #       pdb_string = pdb_file.read()\n","\n"],"metadata":{"id":"OovSvxxOcJyF"},"id":"OovSvxxOcJyF","execution_count":null,"outputs":[]},{"cell_type":"markdown","id":"d5f382dc","metadata":{"id":"d5f382dc"},"source":["### Building the pose\n","We start from the folded polyA peptide stored in a variable called \"start_pose\""]},{"cell_type":"code","execution_count":null,"id":"f0b2d129","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"f0b2d129","outputId":"c3b7c778-51c8-46d6-e53d-8731e8947764"},"outputs":[{"output_type":"stream","name":"stdout","text":["core.import_pose.import_pose: File '/content/google_drive/MyDrive/BIP_24-25/5-Protein_design/mc_final.pdb' automatically determined to be of type PDB\n"]}],"source":["#start_pose = pyrosetta.pose_from_sequence('A' * 10)\n","# Better to start from the folded peptide we obtained earlier\n","\n","scorefxn = pyrosetta.create_score_function(\"ref2015_cart.wts\")\n","start_pose = pose_from_pdb(\"/content/google_drive/MyDrive/BIP_24-25/5-Protein_design/mc_final.pdb\") # Call the function correctly\n","pose = start_pose.clone()\n","pose = start_pose.clone()\n"]},{"cell_type":"code","execution_count":null,"id":"33748443","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"33748443","outputId":"be0d37c8-ad83-4603-8700-9876f9447df6"},"outputs":[{"output_type":"stream","name":"stdout","text":["core.energy_methods.CartesianBondedEnergy: Creating new peptide-bonded energy container (10)\n","basic.io.database: Database file opened: scoring/score_functions/elec_cp_reps.dat\n","core.scoring.elec.util: Read 40 countpair representative atoms\n","core.pack.dunbrack.RotamerLibrary: shapovalov_lib_fixes_enable option is true.\n","core.pack.dunbrack.RotamerLibrary: shapovalov_lib::shap_dun10_smooth_level of 1( aka lowest_smooth ) got activated.\n","core.pack.dunbrack.RotamerLibrary: Binary rotamer library selected: /usr/local/lib/python3.10/dist-packages/pyrosetta/database/rotamer/shapovalov/StpDwn_0-0-0/Dunbrack10.lib.bin\n","core.pack.dunbrack.RotamerLibrary: Using Dunbrack library binary file '/usr/local/lib/python3.10/dist-packages/pyrosetta/database/rotamer/shapovalov/StpDwn_0-0-0/Dunbrack10.lib.bin'.\n","core.pack.dunbrack.RotamerLibrary: Dunbrack 2010 library took 0.445813 seconds to load from binary\n"]},{"output_type":"execute_result","data":{"text/plain":["-5.371127246023434"]},"metadata":{},"execution_count":8}],"source":["# measure energy level\n","scorefxn(pose)"]},{"cell_type":"markdown","id":"0e4733a1","metadata":{"id":"0e4733a1"},"source":["We will use the function above to inspect the initial conformation of the polyA peptide."]},{"cell_type":"code","execution_count":null,"id":"4858b78f","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":617},"id":"4858b78f","outputId":"5e43c38a-dcc1-4459-f2a4-675b985799a1"},"outputs":[{"output_type":"display_data","data":{"application/3dmoljs_load.v0":"<div id=\"3dmolviewer_1726937142863611\"  style=\"position: relative; width: 800px; height: 600px;\">\n        <p id=\"3dmolwarning_1726937142863611\" style=\"background-color:#ffcccc;color:black\">3Dmol.js failed to load for some reason.  Please check your browser console for error messages.<br></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    //this is to ignore the existence of requirejs amd\n    var savedexports, savedmodule;\n    if (typeof exports !== 'undefined') savedexports = exports;\n    else exports = {}\n    if (typeof module !== 'undefined') savedmodule = module;\n    else module = {}\n\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n        exports = savedexports;\n        module = savedmodule;\n        resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://cdnjs.cloudflare.com/ajax/libs/3Dmol/2.4.0/3Dmol-min.js');\n}\n\nvar viewer_1726937142863611 = null;\nvar warn = document.getElementById(\"3dmolwarning_1726937142863611\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_1726937142863611 = $3Dmol.createViewer(document.getElementById(\"3dmolviewer_1726937142863611\"),{backgroundColor:\"white\"});\nviewer_1726937142863611.zoomTo();\n\tviewer_1726937142863611.addModel(\"HEADER                                            21-SEP-24   XXXX              \\nEXPDTA    THEORETICAL MODEL                                                     \\nREMARK 220                                                                      \\nREMARK 220 EXPERIMENTAL DETAILS                                                 \\nREMARK 220  EXPERIMENT TYPE                : THEORETICAL MODELLING              \\nREMARK 220  DATE OF DATA COLLECTION        : 21-SEP-24                          \\nREMARK 220                                                                      \\nREMARK 220 REMARK: MODEL GENERATED BY ROSETTA                                   \\nREMARK 220  VERSION 2024.19+release.a34b73c40f                                  \\nATOM      1  N   ALA A   1       0.557  -0.537   0.388  1.00  0.00           N  \\nATOM      2  CA  ALA A   1       1.951  -0.256   0.067  1.00  0.00           C  \\nATOM      3  C   ALA A   1       2.206   1.231  -0.040  1.00  0.00           C  \\nATOM      4  O   ALA A   1       1.835   1.861  -1.034  1.00  0.00           O  \\nATOM      5  CB  ALA A   1       2.348  -0.914  -1.245  1.00  0.00           C  \\nATOM      6 1H   ALA A   1       0.475  -1.486   0.725  1.00  0.00           H  \\nATOM      7 2H   ALA A   1       0.239   0.101   1.106  1.00  0.00           H  \\nATOM      8 3H   ALA A   1      -0.011  -0.425  -0.438  1.00  0.00           H  \\nATOM      9  HA  ALA A   1       2.571  -0.653   0.874  1.00  0.00           H  \\nATOM     10 1HB  ALA A   1       3.398  -0.709  -1.453  1.00  0.00           H  \\nATOM     11 2HB  ALA A   1       2.197  -1.989  -1.172  1.00  0.00           H  \\nATOM     12 3HB  ALA A   1       1.735  -0.518  -2.048  1.00  0.00           H  \\nATOM     13  N   ALA A   2       2.910   1.791   0.940  1.00  0.00           N  \\nATOM     14  CA  ALA A   2       3.211   3.224   0.916  1.00  0.00           C  \\nATOM     15  C   ALA A   2       3.973   3.583  -0.353  1.00  0.00           C  \\nATOM     16  O   ALA A   2       3.807   4.673  -0.894  1.00  0.00           O  \\nATOM     17  CB  ALA A   2       4.023   3.620   2.133  1.00  0.00           C  \\nATOM     18  H   ALA A   2       3.221   1.240   1.730  1.00  0.00           H  \\nATOM     19  HA  ALA A   2       2.271   3.774   0.919  1.00  0.00           H  \\nATOM     20 1HB  ALA A   2       4.223   4.691   2.101  1.00  0.00           H  \\nATOM     21 2HB  ALA A   2       3.467   3.383   3.040  1.00  0.00           H  \\nATOM     22 3HB  ALA A   2       4.967   3.078   2.134  1.00  0.00           H  \\nATOM     23  N   ALA A   3       4.812   2.662  -0.823  1.00  0.00           N  \\nATOM     24  CA  ALA A   3       5.577   2.882  -2.034  1.00  0.00           C  \\nATOM     25  C   ALA A   3       4.663   3.123  -3.232  1.00  0.00           C  \\nATOM     26  O   ALA A   3       5.004   3.894  -4.132  1.00  0.00           O  \\nATOM     27  CB  ALA A   3       6.459   1.685  -2.311  1.00  0.00           C  \\nATOM     28  H   ALA A   3       4.910   1.790  -0.329  1.00  0.00           H  \\nATOM     29  HA  ALA A   3       6.196   3.767  -1.889  1.00  0.00           H  \\nATOM     30 1HB  ALA A   3       7.047   1.867  -3.209  1.00  0.00           H  \\nATOM     31 2HB  ALA A   3       7.128   1.524  -1.465  1.00  0.00           H  \\nATOM     32 3HB  ALA A   3       5.837   0.805  -2.455  1.00  0.00           H  \\nATOM     33  N   ALA A   4       3.517   2.435  -3.276  1.00  0.00           N  \\nATOM     34  CA  ALA A   4       2.605   2.606  -4.389  1.00  0.00           C  \\nATOM     35  C   ALA A   4       2.035   4.001  -4.338  1.00  0.00           C  \\nATOM     36  O   ALA A   4       1.894   4.672  -5.362  1.00  0.00           O  \\nATOM     37  CB  ALA A   4       1.492   1.581  -4.344  1.00  0.00           C  \\nATOM     38  H   ALA A   4       3.233   1.852  -2.500  1.00  0.00           H  \\nATOM     39  HA  ALA A   4       3.163   2.493  -5.318  1.00  0.00           H  \\nATOM     40 1HB  ALA A   4       0.827   1.730  -5.190  1.00  0.00           H  \\nATOM     41 2HB  ALA A   4       1.920   0.581  -4.387  1.00  0.00           H  \\nATOM     42 3HB  ALA A   4       0.933   1.695  -3.419  1.00  0.00           H  \\nATOM     43  N   ALA A   5       1.716   4.446  -3.120  1.00  0.00           N  \\nATOM     44  CA  ALA A   5       1.175   5.783  -2.944  1.00  0.00           C  \\nATOM     45  C   ALA A   5       2.204   6.816  -3.382  1.00  0.00           C  \\nATOM     46  O   ALA A   5       1.874   7.805  -4.044  1.00  0.00           O  \\nATOM     47  CB  ALA A   5       0.780   6.009  -1.500  1.00  0.00           C  \\nATOM     48  H   ALA A   5       1.855   3.827  -2.317  1.00  0.00           H  \\nATOM     49  HA  ALA A   5       0.294   5.881  -3.577  1.00  0.00           H  \\nATOM     50 1HB  ALA A   5       0.363   7.006  -1.390  1.00  0.00           H  \\nATOM     51 2HB  ALA A   5       0.038   5.267  -1.209  1.00  0.00           H  \\nATOM     52 3HB  ALA A   5       1.651   5.913  -0.863  1.00  0.00           H  \\nATOM     53  N   ALA A   6       3.466   6.565  -3.033  1.00  0.00           N  \\nATOM     54  CA  ALA A   6       4.532   7.468  -3.411  1.00  0.00           C  \\nATOM     55  C   ALA A   6       4.638   7.548  -4.923  1.00  0.00           C  \\nATOM     56  O   ALA A   6       4.789   8.632  -5.488  1.00  0.00           O  \\nATOM     57  CB  ALA A   6       5.847   7.003  -2.821  1.00  0.00           C  \\nATOM     58  H   ALA A   6       3.675   5.742  -2.465  1.00  0.00           H  \\nATOM     59  HA  ALA A   6       4.292   8.459  -3.027  1.00  0.00           H  \\nATOM     60 1HB  ALA A   6       6.636   7.697  -3.095  1.00  0.00           H  \\nATOM     61 2HB  ALA A   6       5.758   6.956  -1.735  1.00  0.00           H  \\nATOM     62 3HB  ALA A   6       6.084   6.017  -3.204  1.00  0.00           H  \\nATOM     63  N   ALA A   7       4.515   6.398  -5.591  1.00  0.00           N  \\nATOM     64  CA  ALA A   7       4.559   6.376  -7.042  1.00  0.00           C  \\nATOM     65  C   ALA A   7       3.386   7.152  -7.628  1.00  0.00           C  \\nATOM     66  O   ALA A   7       3.546   7.924  -8.567  1.00  0.00           O  \\nATOM     67  CB  ALA A   7       4.540   4.944  -7.538  1.00  0.00           C  \\nATOM     68  H   ALA A   7       4.416   5.522  -5.079  1.00  0.00           H  \\nATOM     69  HA  ALA A   7       5.482   6.858  -7.362  1.00  0.00           H  \\nATOM     70 1HB  ALA A   7       4.596   4.934  -8.623  1.00  0.00           H  \\nATOM     71 2HB  ALA A   7       5.391   4.409  -7.119  1.00  0.00           H  \\nATOM     72 3HB  ALA A   7       3.621   4.462  -7.218  1.00  0.00           H  \\nATOM     73  N   ALA A   8       2.203   6.984  -7.036  1.00  0.00           N  \\nATOM     74  CA  ALA A   8       0.994   7.659  -7.495  1.00  0.00           C  \\nATOM     75  C   ALA A   8       1.147   9.173  -7.423  1.00  0.00           C  \\nATOM     76  O   ALA A   8       0.623   9.910  -8.260  1.00  0.00           O  \\nATOM     77  CB  ALA A   8      -0.197   7.219  -6.662  1.00  0.00           C  \\nATOM     78  H   ALA A   8       2.129   6.328  -6.258  1.00  0.00           H  \\nATOM     79  HA  ALA A   8       0.826   7.381  -8.534  1.00  0.00           H  \\nATOM     80 1HB  ALA A   8      -1.097   7.704  -7.029  1.00  0.00           H  \\nATOM     81 2HB  ALA A   8      -0.306   6.137  -6.735  1.00  0.00           H  \\nATOM     82 3HB  ALA A   8      -0.036   7.494  -5.628  1.00  0.00           H  \\nATOM     83  N   ALA A   9       1.879   9.624  -6.411  1.00  0.00           N  \\nATOM     84  CA  ALA A   9       2.140  11.027  -6.140  1.00  0.00           C  \\nATOM     85  C   ALA A   9       3.295  11.618  -6.963  1.00  0.00           C  \\nATOM     86  O   ALA A   9       3.613  12.800  -6.796  1.00  0.00           O  \\nATOM     87  CB  ALA A   9       2.452  11.191  -4.666  1.00  0.00           C  \\nATOM     88  H   ALA A   9       2.247   8.938  -5.749  1.00  0.00           H  \\nATOM     89  HA  ALA A   9       1.238  11.585  -6.385  1.00  0.00           H  \\nATOM     90 1HB  ALA A   9       2.607  12.245  -4.440  1.00  0.00           H  \\nATOM     91 2HB  ALA A   9       1.620  10.809  -4.075  1.00  0.00           H  \\nATOM     92 3HB  ALA A   9       3.354  10.629  -4.431  1.00  0.00           H  \\nATOM     93  N   ALA A  10       3.953  10.804  -7.795  1.00  0.00           N  \\nATOM     94  CA  ALA A  10       5.111  11.243  -8.565  1.00  0.00           C  \\nATOM     95  C   ALA A  10       4.766  11.476 -10.035  1.00  0.00           C  \\nATOM     96  O   ALA A  10       4.493  12.611 -10.430  1.00  0.00           O  \\nATOM     97  OXT ALA A  10       5.149  10.640 -10.854  1.00  0.00           O  \\nATOM     98  CB  ALA A  10       6.210  10.201  -8.445  1.00  0.00           C  \\nATOM     99  H   ALA A  10       3.649   9.836  -7.908  1.00  0.00           H  \\nATOM    100  HA  ALA A  10       5.462  12.183  -8.144  1.00  0.00           H  \\nATOM    101 1HB  ALA A  10       7.092  10.533  -8.989  1.00  0.00           H  \\nATOM    102 2HB  ALA A  10       6.462  10.058  -7.391  1.00  0.00           H  \\nATOM    103 3HB  ALA A  10       5.859   9.255  -8.859  1.00  0.00           H  \\nTER                                                                             \\n# All scores below are weighted scores, not raw scores.\\n#BEGIN_POSE_ENERGIES_TABLE /tmp/tmpnw4gnoq2.pdb\\nlabel fa_atr fa_rep fa_sol fa_intra_rep fa_intra_sol_xover4 lk_ball_wtd fa_elec hbond_sr_bb hbond_lr_bb hbond_bb_sc hbond_sc dslf_fa13 omega fa_dun p_aa_pp yhh_planarity ref rama_prepro cart_bonded total\\nweights 1 0.55 1 0.005 1 1 1 1 1 1 1 1.25 0.4 0.7 0.6 0.625 1 0.45 0.5 NA\\npose -28.7684 2.54769 21.9707 0.01818 0 -0.93584 -9.22838 -6.89712 0 0 0 0 -0.1157 0 -1.69369 0 13.2468 -3.47508 1.64261 -11.6882\\nALA:NtermProteinFull_1 -2.03616 0.28429 1.17 0.00257 0 -0.1008 -0.96274 0 0 0 0 0 -0.06895 0 0 0 1.32468 0 0.08376 -0.30333\\nALA_2 -2.2071 0.23585 1.74228 0.00141 0 -0.02658 -1.15589 0 0 0 0 0 -0.01645 0 -0.39282 0 1.32468 -0.29773 0.10405 -0.68831\\nALA_3 -2.95681 0.34861 1.98203 0.00155 0 -0.11736 -0.4756 0 0 0 0 0 -0.04525 0 -0.35806 0 1.32468 -0.591 0.15651 -0.7307\\nALA_4 -3.73454 0.33097 2.67496 0.00143 0 -0.14114 -1.3273 0 0 0 0 0 0.0293 0 -0.07603 0 1.32468 -0.38209 0.2016 -1.09816\\nALA_5 -3.82579 0.32002 2.93819 0.00138 0 -0.05002 -1.13897 0 0 0 0 0 0.01696 0 -0.1993 0 1.32468 -0.25729 0.12497 -0.74517\\nALA_6 -3.89651 0.24333 2.96441 0.00148 0 -0.03555 -0.96889 0 0 0 0 0 0.06471 0 -0.20279 0 1.32468 -0.3271 0.11774 -0.7145\\nALA_7 -3.55911 0.30312 2.89453 0.00146 0 -0.17654 -0.97521 0 0 0 0 0 -0.01556 0 -0.16143 0 1.32468 -0.2973 0.11208 -0.54928\\nALA_8 -2.6395 0.20694 2.12735 0.00143 0 -0.07489 -0.80878 0 0 0 0 0 -0.0567 0 -0.34131 0 1.32468 -0.45575 0.14603 -0.5705\\nALA_9 -2.18912 0.17661 1.72808 0.00229 0 -0.10559 -1.04973 0 0 0 0 0 -0.02377 0 0.03806 0 1.32468 -0.59196 0.26311 -0.42734\\nALA:CtermProteinFull_10 -1.72379 0.09794 1.74888 0.00319 0 -0.10736 -0.36526 0 0 0 0 0 0 0 0 0 1.32468 -0.27487 0.33277 1.03617\\n#END_POSE_ENERGIES_TABLE /tmp/tmpnw4gnoq2.pdb\\n\\n\",\"pdb\");\n\tviewer_1726937142863611.setStyle({\"stick\": {\"colorscheme\": \"greenCarbon\"}});\n\tviewer_1726937142863611.zoomTo();\nviewer_1726937142863611.render();\n});\n</script>","text/html":["<div id=\"3dmolviewer_1726937142863611\"  style=\"position: relative; width: 800px; height: 600px;\">\n","        <p id=\"3dmolwarning_1726937142863611\" style=\"background-color:#ffcccc;color:black\">3Dmol.js failed to load for some reason.  Please check your browser console for error messages.<br></p>\n","        </div>\n","<script>\n","\n","var loadScriptAsync = function(uri){\n","  return new Promise((resolve, reject) => {\n","    //this is to ignore the existence of requirejs amd\n","    var savedexports, savedmodule;\n","    if (typeof exports !== 'undefined') savedexports = exports;\n","    else exports = {}\n","    if (typeof module !== 'undefined') savedmodule = module;\n","    else module = {}\n","\n","    var tag = document.createElement('script');\n","    tag.src = uri;\n","    tag.async = true;\n","    tag.onload = () => {\n","        exports = savedexports;\n","        module = savedmodule;\n","        resolve();\n","    };\n","  var firstScriptTag = document.getElementsByTagName('script')[0];\n","  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n","});\n","};\n","\n","if(typeof $3Dmolpromise === 'undefined') {\n","$3Dmolpromise = null;\n","  $3Dmolpromise = loadScriptAsync('https://cdnjs.cloudflare.com/ajax/libs/3Dmol/2.4.0/3Dmol-min.js');\n","}\n","\n","var viewer_1726937142863611 = null;\n","var warn = document.getElementById(\"3dmolwarning_1726937142863611\");\n","if(warn) {\n","    warn.parentNode.removeChild(warn);\n","}\n","$3Dmolpromise.then(function() {\n","viewer_1726937142863611 = $3Dmol.createViewer(document.getElementById(\"3dmolviewer_1726937142863611\"),{backgroundColor:\"white\"});\n","viewer_1726937142863611.zoomTo();\n","\tviewer_1726937142863611.addModel(\"HEADER                                            21-SEP-24   XXXX              \\nEXPDTA    THEORETICAL MODEL                                                     \\nREMARK 220                                                                      \\nREMARK 220 EXPERIMENTAL DETAILS                                                 \\nREMARK 220  EXPERIMENT TYPE                : THEORETICAL MODELLING              \\nREMARK 220  DATE OF DATA COLLECTION        : 21-SEP-24                          \\nREMARK 220                                                                      \\nREMARK 220 REMARK: MODEL GENERATED BY ROSETTA                                   \\nREMARK 220  VERSION 2024.19+release.a34b73c40f                                  \\nATOM      1  N   ALA A   1       0.557  -0.537   0.388  1.00  0.00           N  \\nATOM      2  CA  ALA A   1       1.951  -0.256   0.067  1.00  0.00           C  \\nATOM      3  C   ALA A   1       2.206   1.231  -0.040  1.00  0.00           C  \\nATOM      4  O   ALA A   1       1.835   1.861  -1.034  1.00  0.00           O  \\nATOM      5  CB  ALA A   1       2.348  -0.914  -1.245  1.00  0.00           C  \\nATOM      6 1H   ALA A   1       0.475  -1.486   0.725  1.00  0.00           H  \\nATOM      7 2H   ALA A   1       0.239   0.101   1.106  1.00  0.00           H  \\nATOM      8 3H   ALA A   1      -0.011  -0.425  -0.438  1.00  0.00           H  \\nATOM      9  HA  ALA A   1       2.571  -0.653   0.874  1.00  0.00           H  \\nATOM     10 1HB  ALA A   1       3.398  -0.709  -1.453  1.00  0.00           H  \\nATOM     11 2HB  ALA A   1       2.197  -1.989  -1.172  1.00  0.00           H  \\nATOM     12 3HB  ALA A   1       1.735  -0.518  -2.048  1.00  0.00           H  \\nATOM     13  N   ALA A   2       2.910   1.791   0.940  1.00  0.00           N  \\nATOM     14  CA  ALA A   2       3.211   3.224   0.916  1.00  0.00           C  \\nATOM     15  C   ALA A   2       3.973   3.583  -0.353  1.00  0.00           C  \\nATOM     16  O   ALA A   2       3.807   4.673  -0.894  1.00  0.00           O  \\nATOM     17  CB  ALA A   2       4.023   3.620   2.133  1.00  0.00           C  \\nATOM     18  H   ALA A   2       3.221   1.240   1.730  1.00  0.00           H  \\nATOM     19  HA  ALA A   2       2.271   3.774   0.919  1.00  0.00           H  \\nATOM     20 1HB  ALA A   2       4.223   4.691   2.101  1.00  0.00           H  \\nATOM     21 2HB  ALA A   2       3.467   3.383   3.040  1.00  0.00           H  \\nATOM     22 3HB  ALA A   2       4.967   3.078   2.134  1.00  0.00           H  \\nATOM     23  N   ALA A   3       4.812   2.662  -0.823  1.00  0.00           N  \\nATOM     24  CA  ALA A   3       5.577   2.882  -2.034  1.00  0.00           C  \\nATOM     25  C   ALA A   3       4.663   3.123  -3.232  1.00  0.00           C  \\nATOM     26  O   ALA A   3       5.004   3.894  -4.132  1.00  0.00           O  \\nATOM     27  CB  ALA A   3       6.459   1.685  -2.311  1.00  0.00           C  \\nATOM     28  H   ALA A   3       4.910   1.790  -0.329  1.00  0.00           H  \\nATOM     29  HA  ALA A   3       6.196   3.767  -1.889  1.00  0.00           H  \\nATOM     30 1HB  ALA A   3       7.047   1.867  -3.209  1.00  0.00           H  \\nATOM     31 2HB  ALA A   3       7.128   1.524  -1.465  1.00  0.00           H  \\nATOM     32 3HB  ALA A   3       5.837   0.805  -2.455  1.00  0.00           H  \\nATOM     33  N   ALA A   4       3.517   2.435  -3.276  1.00  0.00           N  \\nATOM     34  CA  ALA A   4       2.605   2.606  -4.389  1.00  0.00           C  \\nATOM     35  C   ALA A   4       2.035   4.001  -4.338  1.00  0.00           C  \\nATOM     36  O   ALA A   4       1.894   4.672  -5.362  1.00  0.00           O  \\nATOM     37  CB  ALA A   4       1.492   1.581  -4.344  1.00  0.00           C  \\nATOM     38  H   ALA A   4       3.233   1.852  -2.500  1.00  0.00           H  \\nATOM     39  HA  ALA A   4       3.163   2.493  -5.318  1.00  0.00           H  \\nATOM     40 1HB  ALA A   4       0.827   1.730  -5.190  1.00  0.00           H  \\nATOM     41 2HB  ALA A   4       1.920   0.581  -4.387  1.00  0.00           H  \\nATOM     42 3HB  ALA A   4       0.933   1.695  -3.419  1.00  0.00           H  \\nATOM     43  N   ALA A   5       1.716   4.446  -3.120  1.00  0.00           N  \\nATOM     44  CA  ALA A   5       1.175   5.783  -2.944  1.00  0.00           C  \\nATOM     45  C   ALA A   5       2.204   6.816  -3.382  1.00  0.00           C  \\nATOM     46  O   ALA A   5       1.874   7.805  -4.044  1.00  0.00           O  \\nATOM     47  CB  ALA A   5       0.780   6.009  -1.500  1.00  0.00           C  \\nATOM     48  H   ALA A   5       1.855   3.827  -2.317  1.00  0.00           H  \\nATOM     49  HA  ALA A   5       0.294   5.881  -3.577  1.00  0.00           H  \\nATOM     50 1HB  ALA A   5       0.363   7.006  -1.390  1.00  0.00           H  \\nATOM     51 2HB  ALA A   5       0.038   5.267  -1.209  1.00  0.00           H  \\nATOM     52 3HB  ALA A   5       1.651   5.913  -0.863  1.00  0.00           H  \\nATOM     53  N   ALA A   6       3.466   6.565  -3.033  1.00  0.00           N  \\nATOM     54  CA  ALA A   6       4.532   7.468  -3.411  1.00  0.00           C  \\nATOM     55  C   ALA A   6       4.638   7.548  -4.923  1.00  0.00           C  \\nATOM     56  O   ALA A   6       4.789   8.632  -5.488  1.00  0.00           O  \\nATOM     57  CB  ALA A   6       5.847   7.003  -2.821  1.00  0.00           C  \\nATOM     58  H   ALA A   6       3.675   5.742  -2.465  1.00  0.00           H  \\nATOM     59  HA  ALA A   6       4.292   8.459  -3.027  1.00  0.00           H  \\nATOM     60 1HB  ALA A   6       6.636   7.697  -3.095  1.00  0.00           H  \\nATOM     61 2HB  ALA A   6       5.758   6.956  -1.735  1.00  0.00           H  \\nATOM     62 3HB  ALA A   6       6.084   6.017  -3.204  1.00  0.00           H  \\nATOM     63  N   ALA A   7       4.515   6.398  -5.591  1.00  0.00           N  \\nATOM     64  CA  ALA A   7       4.559   6.376  -7.042  1.00  0.00           C  \\nATOM     65  C   ALA A   7       3.386   7.152  -7.628  1.00  0.00           C  \\nATOM     66  O   ALA A   7       3.546   7.924  -8.567  1.00  0.00           O  \\nATOM     67  CB  ALA A   7       4.540   4.944  -7.538  1.00  0.00           C  \\nATOM     68  H   ALA A   7       4.416   5.522  -5.079  1.00  0.00           H  \\nATOM     69  HA  ALA A   7       5.482   6.858  -7.362  1.00  0.00           H  \\nATOM     70 1HB  ALA A   7       4.596   4.934  -8.623  1.00  0.00           H  \\nATOM     71 2HB  ALA A   7       5.391   4.409  -7.119  1.00  0.00           H  \\nATOM     72 3HB  ALA A   7       3.621   4.462  -7.218  1.00  0.00           H  \\nATOM     73  N   ALA A   8       2.203   6.984  -7.036  1.00  0.00           N  \\nATOM     74  CA  ALA A   8       0.994   7.659  -7.495  1.00  0.00           C  \\nATOM     75  C   ALA A   8       1.147   9.173  -7.423  1.00  0.00           C  \\nATOM     76  O   ALA A   8       0.623   9.910  -8.260  1.00  0.00           O  \\nATOM     77  CB  ALA A   8      -0.197   7.219  -6.662  1.00  0.00           C  \\nATOM     78  H   ALA A   8       2.129   6.328  -6.258  1.00  0.00           H  \\nATOM     79  HA  ALA A   8       0.826   7.381  -8.534  1.00  0.00           H  \\nATOM     80 1HB  ALA A   8      -1.097   7.704  -7.029  1.00  0.00           H  \\nATOM     81 2HB  ALA A   8      -0.306   6.137  -6.735  1.00  0.00           H  \\nATOM     82 3HB  ALA A   8      -0.036   7.494  -5.628  1.00  0.00           H  \\nATOM     83  N   ALA A   9       1.879   9.624  -6.411  1.00  0.00           N  \\nATOM     84  CA  ALA A   9       2.140  11.027  -6.140  1.00  0.00           C  \\nATOM     85  C   ALA A   9       3.295  11.618  -6.963  1.00  0.00           C  \\nATOM     86  O   ALA A   9       3.613  12.800  -6.796  1.00  0.00           O  \\nATOM     87  CB  ALA A   9       2.452  11.191  -4.666  1.00  0.00           C  \\nATOM     88  H   ALA A   9       2.247   8.938  -5.749  1.00  0.00           H  \\nATOM     89  HA  ALA A   9       1.238  11.585  -6.385  1.00  0.00           H  \\nATOM     90 1HB  ALA A   9       2.607  12.245  -4.440  1.00  0.00           H  \\nATOM     91 2HB  ALA A   9       1.620  10.809  -4.075  1.00  0.00           H  \\nATOM     92 3HB  ALA A   9       3.354  10.629  -4.431  1.00  0.00           H  \\nATOM     93  N   ALA A  10       3.953  10.804  -7.795  1.00  0.00           N  \\nATOM     94  CA  ALA A  10       5.111  11.243  -8.565  1.00  0.00           C  \\nATOM     95  C   ALA A  10       4.766  11.476 -10.035  1.00  0.00           C  \\nATOM     96  O   ALA A  10       4.493  12.611 -10.430  1.00  0.00           O  \\nATOM     97  OXT ALA A  10       5.149  10.640 -10.854  1.00  0.00           O  \\nATOM     98  CB  ALA A  10       6.210  10.201  -8.445  1.00  0.00           C  \\nATOM     99  H   ALA A  10       3.649   9.836  -7.908  1.00  0.00           H  \\nATOM    100  HA  ALA A  10       5.462  12.183  -8.144  1.00  0.00           H  \\nATOM    101 1HB  ALA A  10       7.092  10.533  -8.989  1.00  0.00           H  \\nATOM    102 2HB  ALA A  10       6.462  10.058  -7.391  1.00  0.00           H  \\nATOM    103 3HB  ALA A  10       5.859   9.255  -8.859  1.00  0.00           H  \\nTER                                                                             \\n# All scores below are weighted scores, not raw scores.\\n#BEGIN_POSE_ENERGIES_TABLE /tmp/tmpnw4gnoq2.pdb\\nlabel fa_atr fa_rep fa_sol fa_intra_rep fa_intra_sol_xover4 lk_ball_wtd fa_elec hbond_sr_bb hbond_lr_bb hbond_bb_sc hbond_sc dslf_fa13 omega fa_dun p_aa_pp yhh_planarity ref rama_prepro cart_bonded total\\nweights 1 0.55 1 0.005 1 1 1 1 1 1 1 1.25 0.4 0.7 0.6 0.625 1 0.45 0.5 NA\\npose -28.7684 2.54769 21.9707 0.01818 0 -0.93584 -9.22838 -6.89712 0 0 0 0 -0.1157 0 -1.69369 0 13.2468 -3.47508 1.64261 -11.6882\\nALA:NtermProteinFull_1 -2.03616 0.28429 1.17 0.00257 0 -0.1008 -0.96274 0 0 0 0 0 -0.06895 0 0 0 1.32468 0 0.08376 -0.30333\\nALA_2 -2.2071 0.23585 1.74228 0.00141 0 -0.02658 -1.15589 0 0 0 0 0 -0.01645 0 -0.39282 0 1.32468 -0.29773 0.10405 -0.68831\\nALA_3 -2.95681 0.34861 1.98203 0.00155 0 -0.11736 -0.4756 0 0 0 0 0 -0.04525 0 -0.35806 0 1.32468 -0.591 0.15651 -0.7307\\nALA_4 -3.73454 0.33097 2.67496 0.00143 0 -0.14114 -1.3273 0 0 0 0 0 0.0293 0 -0.07603 0 1.32468 -0.38209 0.2016 -1.09816\\nALA_5 -3.82579 0.32002 2.93819 0.00138 0 -0.05002 -1.13897 0 0 0 0 0 0.01696 0 -0.1993 0 1.32468 -0.25729 0.12497 -0.74517\\nALA_6 -3.89651 0.24333 2.96441 0.00148 0 -0.03555 -0.96889 0 0 0 0 0 0.06471 0 -0.20279 0 1.32468 -0.3271 0.11774 -0.7145\\nALA_7 -3.55911 0.30312 2.89453 0.00146 0 -0.17654 -0.97521 0 0 0 0 0 -0.01556 0 -0.16143 0 1.32468 -0.2973 0.11208 -0.54928\\nALA_8 -2.6395 0.20694 2.12735 0.00143 0 -0.07489 -0.80878 0 0 0 0 0 -0.0567 0 -0.34131 0 1.32468 -0.45575 0.14603 -0.5705\\nALA_9 -2.18912 0.17661 1.72808 0.00229 0 -0.10559 -1.04973 0 0 0 0 0 -0.02377 0 0.03806 0 1.32468 -0.59196 0.26311 -0.42734\\nALA:CtermProteinFull_10 -1.72379 0.09794 1.74888 0.00319 0 -0.10736 -0.36526 0 0 0 0 0 0 0 0 0 1.32468 -0.27487 0.33277 1.03617\\n#END_POSE_ENERGIES_TABLE /tmp/tmpnw4gnoq2.pdb\\n\\n\",\"pdb\");\n","\tviewer_1726937142863611.setStyle({\"stick\": {\"colorscheme\": \"greenCarbon\"}});\n","\tviewer_1726937142863611.zoomTo();\n","viewer_1726937142863611.render();\n","});\n","</script>"]},"metadata":{}}],"source":["display_protein(pose)"]},{"cell_type":"markdown","id":"832bb2d5","metadata":{"id":"832bb2d5"},"source":["### Insert mutations"]},{"cell_type":"markdown","id":"ef3fbe45","metadata":{"id":"ef3fbe45"},"source":["Define the mutations you want to introduce. We need to write them in a resfile.\n","\n","In PyRosetta, a resfile is used to specify which residues should be mutated and to what amino acid they should be changed. The resfile format allows for a lot of flexibility and control over the design process.\n","\n","Here is an example of a resfile that would mutate positions 3 and 7 of a 10 amino acid polyalanine peptide. In this example, we will mutate position 3 to aspartate (D) and position 7 to leucine (L).\n","\n","```\n","NATRO\n","start\n","\n","3 A PIKAA D\n","7 A PIKAA L\n","```\n","\n","* NATRO: This command specifies that all residues not mentioned explicitly should not be repacked or mutated (they remain \"naturally\" as they are).\n","* start: This indicates the beginning of the residue-specific commands.\n","* 3 A PIKAA D: This line tells PyRosetta to mutate residue number 3 in chain A to aspartate. PIKAA stands for \"point mutation using the indicated amino acid.\"\n","* 7 A PIKAA L: This line tells PyRosetta to mutate residue number 7 in chain A to leucine.\n","\n","Check out the design proposed in the `design.resfile`."]},{"cell_type":"code","execution_count":null,"id":"6f1cb47b","metadata":{"id":"6f1cb47b"},"outputs":[],"source":["resfile = \"/content/google_drive/MyDrive/BIP_24-25/5-Protein_design/design.resfile\""]},{"cell_type":"markdown","id":"b62ca769","metadata":{"id":"b62ca769"},"source":["Now we can setup the TaskOperations for the `FastRelax` mover. These tell `FastRelax` which residues to design or repack during the packer steps in `FastRelax`.\n"]},{"cell_type":"markdown","source":["In PyRosetta, **TaskOperations** and **FastRelax** are crucial for controlling how the protein structure is modified, especially during tasks like structure optimization and repacking.\n","\n","\n","\n","**TaskOperations** define specific rules and operations that guide how residues in a protein are manipulated during structure refinement, repacking, and design processes. They specify which residues are allowed to be mutated, which rotamers to sample, and how packing should proceed. In your code:\n","\n","- `TaskFactory`: This object collects all the **TaskOperations**. It's like a manager that bundles all the operations you want to apply to your structure.\n","  \n","- `InitializeFromCommandline()`: This operation sets up the packing task based on any command-line options provided during PyRosetta initialization. It helps initialize the task with default settings.\n","\n","- `IncludeCurrent()`: This operation ensures that the current rotamers (the conformations of side chains) of residues are included in the packing process. It ensures that the current structure is among the options when PyRosetta evaluates side-chain conformations.\n","\n","- `ReadResfile(resfile)`: This reads a **resfile**, a text file that specifies custom rules about which residues should be repacked or redesigned. You can use it to fine-tune how specific regions of the protein are handled.\n","\n","By using these **TaskOperations**, you can define which residues in your protein should be flexible, which ones should remain fixed, or even guide specific mutation and repacking strategies. The `packer_task` created from this allows you to inspect the task and see how it's being applied to your protein (`pose`).\n","\n","\n","\n","**FastRelax** is a powerful mover in PyRosetta designed for fast and efficient protein structure refinement. It combines multiple cycles of energy minimization and side-chain repacking to produce low-energy protein models. It’s a faster variant of the standard **Relax** protocol, but it still provides high-quality results.\n","\n","- **Minimization**: FastRelax performs gradient-based energy minimization, adjusting the protein's backbone and side chains to lower the overall energy.\n","  \n","- **Repacking**: During FastRelax, side chains are repacked to explore different rotamer conformations. This is where **TaskOperations** come into play: they control which residues get repacked, how they are sampled, and what restrictions (e.g., from a resfile) should be applied.\n","\n","FastRelax typically works in several cycles, alternating between repacking and minimizing to drive the protein structure to a local energy minimum.\n","\n","In the code below, the **TaskFactory** you set up defines the rules for how the residues will be repacked during the **FastRelax** protocol, ensuring that the current rotamers are considered and that any custom residue rules from the resfile are respected.\n","\n","\n","The **PackerTask** in PyRosetta defines the specific set of operations and rules used during **repacking** or **design** of side chains in a protein structure. It's essentially a blueprint that controls how the side chains (rotamers) of residues in a protein will be sampled, repacked, or redesigned.\n","\n","##### Key aspects of a **PackerTask**:\n","\n","1. **Repacking**: The PackerTask dictates which residues' side chains can be repacked, meaning their rotamer conformations are optimized while keeping the amino acid identity fixed. This is useful for refining the structure without changing its overall sequence.\n","\n","2. **Design**: In addition to repacking, a PackerTask can also be used to redesign certain residues, meaning the side chain can change to a different amino acid entirely. The rules for this can be specified by the resfile or other task operations.\n","\n","3. **TaskOperations**: A PackerTask is generated by applying **TaskOperations**, which control:\n","   - Which residues are allowed to be modified (packed, repacked, or designed).\n","   - Whether the current side chain conformation is included in the repacking process.\n","   - Any restrictions on amino acid substitutions or packing regions.\n","\n","4. **Residue-level control**: For each residue in the protein, the PackerTask can specify whether it is allowed to move, which rotamers can be sampled, and whether the side chain should be repacked, designed, or remain fixed.\n","\n","\n","In summary, the **PackerTask** is a detailed set of instructions that defines how side chains are handled during repacking or design, ensuring that the specified residues follow the rules set by the task operations and resfile during processes like FastRelax.\n"],"metadata":{"id":"WLP8vEehfhMu"},"id":"WLP8vEehfhMu"},{"cell_type":"code","execution_count":null,"id":"b0d148d1","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"b0d148d1","outputId":"951fc3ce-13d0-4cc8-9e66-530b2cc4ea39"},"outputs":[{"output_type":"stream","name":"stdout","text":["core.pack.task: Packer task: initialize from command line()\n","#Packer_Task\n","\n","Threads to request: ALL AVAILABLE\n","\n","resid\tpack?\tdesign?\tallowed_aas\n","1\tTRUE\tFALSE\tALA:NtermProteinFull\n","2\tTRUE\tFALSE\tALA\n","3\tTRUE\tFALSE\tALA\n","4\tTRUE\tFALSE\tALA\n","5\tTRUE\tTRUE\tPRO\n","6\tTRUE\tFALSE\tALA\n","7\tTRUE\tTRUE\tLYS\n","8\tTRUE\tFALSE\tALA\n","9\tTRUE\tFALSE\tALA\n","10\tTRUE\tFALSE\tALA:CtermProteinFull\n","\n"]}],"source":["# The task factory accepts all the task operations\n","tf = pyrosetta.rosetta.core.pack.task.TaskFactory()\n","\n","# These are pretty standard\n","tf.push_back(pyrosetta.rosetta.core.pack.task.operation.InitializeFromCommandline())\n","tf.push_back(pyrosetta.rosetta.core.pack.task.operation.IncludeCurrent())\n","\n","\n","# Include the resfile\n","tf.push_back(pyrosetta.rosetta.core.pack.task.operation.ReadResfile(resfile))\n","\n","# Convert the task factory into a PackerTask to take a look at it\n","packer_task = tf.create_task_and_apply_taskoperations(pose)\n","# View the PackerTask\n","print(packer_task)"]},{"cell_type":"markdown","id":"ae8332b2","metadata":{"id":"ae8332b2"},"source":[" The PackerTask looks as intended!\n","\n"," Now we can set up a `MoveMap` to specify which torsions are free to minimize during the minimization steps of the `FastDesign` mover.\n","\n"," A `MoveMap` in PyRosetta is a data structure that defines which degrees of freedom (backbone, side chains, or rigid-body jumps) in a protein structure can be moved or optimized during refinement or minimization processes. It's used to control which parts of the protein are flexible and which are fixed during various protocols such as **FastRelax**, **Minimization**, and **CartesianMinimization**.\n","\n","##### Components of a **MoveMap**:\n","\n","1. **Backbone flexibility (bb)**: The protein's backbone consists of the φ (phi) and ψ (psi) torsion angles for each residue. Setting the backbone to be flexible means these torsion angles can change during the optimization.\n","   \n","2. **Side-chain flexibility (chi)**: The side-chain dihedral angles (χ angles) can also be adjusted. Allowing chi flexibility means that side-chain conformations (rotamers) can be optimized during refinement.\n","\n","3. **Jump flexibility (jump)**: Proteins or complexes that contain multiple domains or chains connected by rigid-body jumps (rigid-body transformations between two parts of a structure) can have their relative orientations changed if jumps are flexible.\n","\n","### Breakdown of the code:\n","\n","```python\n","mm = pyrosetta.rosetta.core.kinematics.MoveMap()\n","mm.set_bb(True)   # Allow backbone torsions (phi, psi) to move\n","mm.set_chi(True)  # Allow side-chain torsions (chi) to move\n","mm.set_jump(True) # Allow rigid-body jumps to move\n","```\n","\n","- `mm.set_bb(True)`: This command allows the **backbone** torsions of the protein to move during optimization. If set to `False`, the backbone remains fixed.\n","  \n","- `mm.set_chi(True)`: This enables movement in the **side-chain dihedral angles**. If set to `False`, the side chains remain fixed.\n","\n","- `mm.set_jump(True)`: This allows **rigid-body jumps** (which define the relative positioning of chains or domains) to be altered.\n","\n","##### How the **MoveMap** is used:\n","\n","\n","In the code below, you are allowing **all parts of the protein**—the backbone, side chains, and jumps—to move, making the entire structure flexible during whatever optimization protocol you're running.\n"]},{"cell_type":"code","execution_count":null,"id":"51051a95","metadata":{"id":"51051a95"},"outputs":[],"source":["mm = pyrosetta.rosetta.core.kinematics.MoveMap()\n","mm.set_bb(True)\n","mm.set_chi(True)\n","mm.set_jump(True)"]},{"cell_type":"markdown","id":"2ac89f0c","metadata":{"id":"2ac89f0c"},"source":["Set up `FastDesign`"]},{"cell_type":"code","execution_count":null,"id":"43d83078","metadata":{"id":"43d83078"},"outputs":[],"source":["rel_design = pyrosetta.rosetta.protocols.relax.FastRelax(scorefxn_in=scorefxn, standard_repeats=1, script_file=\"MonomerDesign2019\")\n","rel_design.cartesian(True)\n","rel_design.set_task_factory(tf)\n","rel_design.set_movemap(mm)\n","rel_design.minimize_bond_angles(True)\n","rel_design.minimize_bond_lengths(True)"]},{"cell_type":"markdown","id":"5b604d01","metadata":{"id":"5b604d01"},"source":[" Run the `FastDesign` mover. This is fast"]},{"cell_type":"code","execution_count":null,"id":"737ed6c0","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"737ed6c0","outputId":"28d9616a-b2b6-492a-cc20-053f2950c9c0"},"outputs":[{"output_type":"stream","name":"stdout","text":["core.energy_methods.CartesianBondedEnergy: Creating new peptide-bonded energy container (10)\n","protocols.relax.FastRelax: CMD: repeat  -5.37113  0  0  0.55\n","protocols.relax.FastRelax: CMD: coord_cst_weight  -5.37113  0  0  0.55\n","protocols.relax.FastRelax: CMD: scale:fa_rep  -7.87213  0  0  0.03245\n","core.pack.task: Packer task: initialize from command line()\n","core.pack.pack_rotamers: built 30 rotamers at 10 positions.\n","core.pack.interaction_graph.interaction_graph_factory: Instantiating PDInteractionGraph\n","protocols.relax.FastRelax: CMD: repack  5.69158  0  0  0.03245\n","protocols.relax.FastRelax: CMD: scale:fa_rep  12.7067  0  0  0.0506\n","protocols.relax.FastRelax: CMD: min  -13.5998  0.447376  0.447376  0.0506\n","protocols.relax.FastRelax: CMD: coord_cst_weight  -13.5998  0.447376  0.447376  0.0506\n","protocols.relax.FastRelax: CMD: scale:fa_rep  -10.351  0.447376  0.447376  0.154\n","core.pack.task: Packer task: initialize from command line()\n","core.pack.pack_rotamers: built 33 rotamers at 10 positions.\n","core.pack.interaction_graph.interaction_graph_factory: Instantiating PDInteractionGraph\n","protocols.relax.FastRelax: CMD: repack  -10.351  0.447376  0.447376  0.154\n","protocols.relax.FastRelax: CMD: scale:fa_rep  -9.60787  0.447376  0.447376  0.17765\n","protocols.relax.FastRelax: CMD: min  -11.0499  0.464482  0.464482  0.17765\n","protocols.relax.FastRelax: CMD: coord_cst_weight  -11.0499  0.464482  0.464482  0.17765\n","protocols.relax.FastRelax: CMD: scale:fa_rep  -9.06938  0.464482  0.464482  0.3124\n","core.pack.task: Packer task: initialize from command line()\n","core.pack.pack_rotamers: built 32 rotamers at 10 positions.\n","core.pack.interaction_graph.interaction_graph_factory: Instantiating PDInteractionGraph\n","protocols.relax.FastRelax: CMD: repack  -9.06938  0.464482  0.464482  0.3124\n","protocols.relax.FastRelax: CMD: scale:fa_rep  -8.54394  0.464482  0.464482  0.34815\n","protocols.relax.FastRelax: CMD: min  -9.03814  0.479905  0.479905  0.34815\n","protocols.relax.FastRelax: CMD: coord_cst_weight  -9.03814  0.479905  0.479905  0.34815\n","protocols.relax.FastRelax: CMD: scale:fa_rep  -7.03777  0.479905  0.479905  0.55\n","core.pack.task: Packer task: initialize from command line()\n","core.pack.pack_rotamers: built 32 rotamers at 10 positions.\n","core.pack.interaction_graph.interaction_graph_factory: Instantiating PDInteractionGraph\n","protocols.relax.FastRelax: CMD: repack  -7.03777  0.479905  0.479905  0.55\n","protocols.relax.FastRelax: CMD: min  -7.34381  0.493649  0.493649  0.55\n","protocols.relax.FastRelax: MRP: 0  -7.34381  -7.34381  0.493649  0.493649\n","protocols.relax.FastRelax: CMD: accept_to_best  -7.34381  0.493649  0.493649  0.55\n","protocols.relax.FastRelax: CMD: endrepeat  -7.34381  0.493649  0.493649  0.55\n","protocols::checkpoint: Deleting checkpoints of FastRelax\n","CPU times: user 966 ms, sys: 18 ms, total: 984 ms\n","Wall time: 1 s\n"]}],"source":["d3l7_pose = pose.clone()  # Create a copy of the original pose\n","%time rel_design.apply(d3l7_pose)"]},{"cell_type":"markdown","id":"1e52bf3d","metadata":{"id":"1e52bf3d"},"source":["We can see the result of the design."]},{"cell_type":"code","source":["display_protein(d3l7_pose)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":617},"id":"Yh-MM7Oubxxt","outputId":"4c5b9e44-3537-4a38-ab01-701cd71032eb"},"id":"Yh-MM7Oubxxt","execution_count":null,"outputs":[{"output_type":"display_data","data":{"application/3dmoljs_load.v0":"<div id=\"3dmolviewer_17269384987144978\"  style=\"position: relative; width: 800px; height: 600px;\">\n        <p id=\"3dmolwarning_17269384987144978\" style=\"background-color:#ffcccc;color:black\">3Dmol.js failed to load for some reason.  Please check your browser console for error messages.<br></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    //this is to ignore the existence of requirejs amd\n    var savedexports, savedmodule;\n    if (typeof exports !== 'undefined') savedexports = exports;\n    else exports = {}\n    if (typeof module !== 'undefined') savedmodule = module;\n    else module = {}\n\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n        exports = savedexports;\n        module = savedmodule;\n        resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://cdnjs.cloudflare.com/ajax/libs/3Dmol/2.4.0/3Dmol-min.js');\n}\n\nvar viewer_17269384987144978 = null;\nvar warn = document.getElementById(\"3dmolwarning_17269384987144978\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_17269384987144978 = $3Dmol.createViewer(document.getElementById(\"3dmolviewer_17269384987144978\"),{backgroundColor:\"white\"});\nviewer_17269384987144978.zoomTo();\n\tviewer_17269384987144978.addModel(\"HEADER                                            21-SEP-24   XXXX              \\nEXPDTA    THEORETICAL MODEL                                                     \\nREMARK 220                                                                      \\nREMARK 220 EXPERIMENTAL DETAILS                                                 \\nREMARK 220  EXPERIMENT TYPE                : THEORETICAL MODELLING              \\nREMARK 220  DATE OF DATA COLLECTION        : 21-SEP-24                          \\nREMARK 220                                                                      \\nREMARK 220 REMARK: MODEL GENERATED BY ROSETTA                                   \\nREMARK 220  VERSION 2024.19+release.a34b73c40f                                  \\nATOM      1  N   ALA A   1       1.225  -1.236   0.557  1.00  0.00           N  \\nATOM      2  CA  ALA A   1       2.525  -0.732   0.129  1.00  0.00           C  \\nATOM      3  C   ALA A   1       2.543   0.778   0.102  1.00  0.00           C  \\nATOM      4  O   ALA A   1       1.984   1.385  -0.805  1.00  0.00           O  \\nATOM      5  CB  ALA A   1       2.879  -1.244  -1.257  1.00  0.00           C  \\nATOM      6 1H   ALA A   1       1.301  -2.219   0.775  1.00  0.00           H  \\nATOM      7 2H   ALA A   1       0.922  -0.734   1.382  1.00  0.00           H  \\nATOM      8 3H   ALA A   1       0.549  -1.106  -0.180  1.00  0.00           H  \\nATOM      9  HA  ALA A   1       3.274  -1.073   0.845  1.00  0.00           H  \\nATOM     10 1HB  ALA A   1       3.861  -0.867  -1.543  1.00  0.00           H  \\nATOM     11 2HB  ALA A   1       2.895  -2.333  -1.251  1.00  0.00           H  \\nATOM     12 3HB  ALA A   1       2.135  -0.897  -1.972  1.00  0.00           H  \\nATOM     13  N   ALA A   2       3.255   1.396   1.049  1.00  0.00           N  \\nATOM     14  CA  ALA A   2       3.368   2.865   1.116  1.00  0.00           C  \\nATOM     15  C   ALA A   2       3.969   3.415  -0.177  1.00  0.00           C  \\nATOM     16  O   ALA A   2       3.699   4.549  -0.596  1.00  0.00           O  \\nATOM     17  CB  ALA A   2       4.223   3.279   2.297  1.00  0.00           C  \\nATOM     18  H   ALA A   2       3.718   0.850   1.766  1.00  0.00           H  \\nATOM     19  HA  ALA A   2       2.366   3.283   1.231  1.00  0.00           H  \\nATOM     20 1HB  ALA A   2       4.278   4.364   2.340  1.00  0.00           H  \\nATOM     21 2HB  ALA A   2       3.779   2.900   3.217  1.00  0.00           H  \\nATOM     22 3HB  ALA A   2       5.224   2.869   2.180  1.00  0.00           H  \\nATOM     23  N   ALA A   3       4.811   2.601  -0.796  1.00  0.00           N  \\nATOM     24  CA  ALA A   3       5.449   2.944  -2.043  1.00  0.00           C  \\nATOM     25  C   ALA A   3       4.427   3.209  -3.139  1.00  0.00           C  \\nATOM     26  O   ALA A   3       4.688   4.001  -4.041  1.00  0.00           O  \\nATOM     27  CB  ALA A   3       6.366   1.824  -2.485  1.00  0.00           C  \\nATOM     28  H   ALA A   3       4.999   1.701  -0.386  1.00  0.00           H  \\nATOM     29  HA  ALA A   3       6.034   3.847  -1.886  1.00  0.00           H  \\nATOM     30 1HB  ALA A   3       6.858   2.105  -3.415  1.00  0.00           H  \\nATOM     31 2HB  ALA A   3       7.115   1.642  -1.717  1.00  0.00           H  \\nATOM     32 3HB  ALA A   3       5.780   0.920  -2.643  1.00  0.00           H  \\nATOM     33  N   ALA A   4       3.289   2.514  -3.126  1.00  0.00           N  \\nATOM     34  CA  ALA A   4       2.352   2.715  -4.210  1.00  0.00           C  \\nATOM     35  C   ALA A   4       1.868   4.188  -4.230  1.00  0.00           C  \\nATOM     36  O   ALA A   4       1.968   4.807  -5.293  1.00  0.00           O  \\nATOM     37  CB  ALA A   4       1.251   1.655  -4.157  1.00  0.00           C  \\nATOM     38  H   ALA A   4       3.045   1.903  -2.352  1.00  0.00           H  \\nATOM     39  HA  ALA A   4       2.882   2.560  -5.151  1.00  0.00           H  \\nATOM     40 1HB  ALA A   4       0.587   1.783  -5.010  1.00  0.00           H  \\nATOM     41 2HB  ALA A   4       1.728   0.676  -4.217  1.00  0.00           H  \\nATOM     42 3HB  ALA A   4       0.683   1.671  -3.274  1.00  0.00           H  \\nATOM     43  N   PRO A   5       1.306   4.798  -3.147  1.00  0.00           N  \\nATOM     44  CA  PRO A   5       1.035   6.225  -3.068  1.00  0.00           C  \\nATOM     45  C   PRO A   5       2.239   7.091  -3.403  1.00  0.00           C  \\nATOM     46  O   PRO A   5       2.097   8.134  -4.045  1.00  0.00           O  \\nATOM     47  CB  PRO A   5       0.652   6.400  -1.603  1.00  0.00           C  \\nATOM     48  CG  PRO A   5       0.020   5.103  -1.232  1.00  0.00           C  \\nATOM     49  CD  PRO A   5       0.799   4.058  -1.977  1.00  0.00           C  \\nATOM     50  HA  PRO A   5       0.185   6.463  -3.720  1.00  0.00           H  \\nATOM     51 1HB  PRO A   5       1.554   6.613  -1.005  1.00  0.00           H  \\nATOM     52 2HB  PRO A   5      -0.023   7.260  -1.488  1.00  0.00           H  \\nATOM     53 1HG  PRO A   5       0.080   4.956  -0.144  1.00  0.00           H  \\nATOM     54 2HG  PRO A   5      -1.045   5.107  -1.495  1.00  0.00           H  \\nATOM     55 1HD  PRO A   5       1.575   3.674  -1.376  1.00  0.00           H  \\nATOM     56 2HD  PRO A   5       0.071   3.323  -2.242  1.00  0.00           H  \\nATOM     57  N   ALA A   6       3.445   6.652  -3.005  1.00  0.00           N  \\nATOM     58  CA  ALA A   6       4.624   7.455  -3.306  1.00  0.00           C  \\nATOM     59  C   ALA A   6       4.821   7.585  -4.807  1.00  0.00           C  \\nATOM     60  O   ALA A   6       5.154   8.656  -5.315  1.00  0.00           O  \\nATOM     61  CB  ALA A   6       5.860   6.841  -2.676  1.00  0.00           C  \\nATOM     62  H   ALA A   6       3.546   5.799  -2.449  1.00  0.00           H  \\nATOM     63  HA  ALA A   6       4.468   8.452  -2.891  1.00  0.00           H  \\nATOM     64 1HB  ALA A   6       6.722   7.468  -2.890  1.00  0.00           H  \\nATOM     65 2HB  ALA A   6       5.717   6.763  -1.599  1.00  0.00           H  \\nATOM     66 3HB  ALA A   6       6.023   5.864  -3.093  1.00  0.00           H  \\nATOM     67  N   LYS A   7       4.590   6.485  -5.518  1.00  0.00           N  \\nATOM     68  CA  LYS A   7       4.719   6.468  -6.963  1.00  0.00           C  \\nATOM     69  C   LYS A   7       3.554   7.199  -7.613  1.00  0.00           C  \\nATOM     70  O   LYS A   7       3.739   7.951  -8.567  1.00  0.00           O  \\nATOM     71  CB  LYS A   7       4.805   5.027  -7.453  1.00  0.00           C  \\nATOM     72  CG  LYS A   7       6.112   4.323  -7.071  1.00  0.00           C  \\nATOM     73  CD  LYS A   7       6.132   2.877  -7.525  1.00  0.00           C  \\nATOM     74  CE  LYS A   7       7.429   2.190  -7.118  1.00  0.00           C  \\nATOM     75  NZ  LYS A   7       7.448   0.760  -7.530  1.00  0.00           N  \\nATOM     76  H   LYS A   7       4.341   5.623  -5.030  1.00  0.00           H  \\nATOM     77  HA  LYS A   7       5.639   6.986  -7.234  1.00  0.00           H  \\nATOM     78 1HB  LYS A   7       3.982   4.451  -7.021  1.00  0.00           H  \\nATOM     79 2HB  LYS A   7       4.696   4.996  -8.536  1.00  0.00           H  \\nATOM     80 1HG  LYS A   7       6.950   4.850  -7.526  1.00  0.00           H  \\nATOM     81 2HG  LYS A   7       6.233   4.358  -5.991  1.00  0.00           H  \\nATOM     82 1HD  LYS A   7       5.294   2.351  -7.061  1.00  0.00           H  \\nATOM     83 2HD  LYS A   7       6.019   2.827  -8.606  1.00  0.00           H  \\nATOM     84 1HE  LYS A   7       8.266   2.705  -7.587  1.00  0.00           H  \\nATOM     85 2HE  LYS A   7       7.538   2.249  -6.035  1.00  0.00           H  \\nATOM     86 1HZ  LYS A   7       8.321   0.338  -7.244  1.00  0.00           H  \\nATOM     87 2HZ  LYS A   7       6.679   0.275  -7.089  1.00  0.00           H  \\nATOM     88 3HZ  LYS A   7       7.357   0.696  -8.533  1.00  0.00           H  \\nATOM     89  N   ALA A   8       2.354   7.020  -7.065  1.00  0.00           N  \\nATOM     90  CA  ALA A   8       1.147   7.646  -7.599  1.00  0.00           C  \\nATOM     91  C   ALA A   8       1.260   9.164  -7.597  1.00  0.00           C  \\nATOM     92  O   ALA A   8       0.762   9.847  -8.492  1.00  0.00           O  \\nATOM     93  CB  ALA A   8      -0.059   7.231  -6.780  1.00  0.00           C  \\nATOM     94  H   ALA A   8       2.262   6.387  -6.270  1.00  0.00           H  \\nATOM     95  HA  ALA A   8       1.022   7.312  -8.629  1.00  0.00           H  \\nATOM     96 1HB  ALA A   8      -0.955   7.680  -7.201  1.00  0.00           H  \\nATOM     97 2HB  ALA A   8      -0.149   6.145  -6.792  1.00  0.00           H  \\nATOM     98 3HB  ALA A   8       0.075   7.574  -5.763  1.00  0.00           H  \\nATOM     99  N   ALA A   9       1.934   9.685  -6.577  1.00  0.00           N  \\nATOM    100  CA  ALA A   9       2.145  11.106  -6.364  1.00  0.00           C  \\nATOM    101  C   ALA A   9       3.315  11.694  -7.167  1.00  0.00           C  \\nATOM    102  O   ALA A   9       3.591  12.890  -7.037  1.00  0.00           O  \\nATOM    103  CB  ALA A   9       2.399  11.343  -4.889  1.00  0.00           C  \\nATOM    104  H   ALA A   9       2.288   9.050  -5.861  1.00  0.00           H  \\nATOM    105  HA  ALA A   9       1.237  11.626  -6.666  1.00  0.00           H  \\nATOM    106 1HB  ALA A   9       2.520  12.407  -4.702  1.00  0.00           H  \\nATOM    107 2HB  ALA A   9       1.558  10.961  -4.309  1.00  0.00           H  \\nATOM    108 3HB  ALA A   9       3.309  10.811  -4.598  1.00  0.00           H  \\nATOM    109  N   ALA A  10       4.025  10.864  -7.937  1.00  0.00           N  \\nATOM    110  CA  ALA A  10       5.200  11.305  -8.675  1.00  0.00           C  \\nATOM    111  C   ALA A  10       4.915  11.470 -10.169  1.00  0.00           C  \\nATOM    112  O   ALA A  10       4.627  12.579 -10.624  1.00  0.00           O  \\nATOM    113  OXT ALA A  10       5.351  10.611 -10.936  1.00  0.00           O  \\nATOM    114  CB  ALA A  10       6.324  10.302  -8.462  1.00  0.00           C  \\nATOM    115  H   ALA A  10       3.750   9.886  -8.025  1.00  0.00           H  \\nATOM    116  HA  ALA A  10       5.507  12.273  -8.282  1.00  0.00           H  \\nATOM    117 1HB  ALA A  10       7.220  10.636  -8.982  1.00  0.00           H  \\nATOM    118 2HB  ALA A  10       6.532  10.213  -7.394  1.00  0.00           H  \\nATOM    119 3HB  ALA A  10       6.017   9.330  -8.851  1.00  0.00           H  \\nTER                                                                             \\n# All scores below are weighted scores, not raw scores.\\n#BEGIN_POSE_ENERGIES_TABLE /tmp/tmp1bqdqyrg.pdb\\nlabel fa_atr fa_rep fa_sol fa_intra_rep fa_intra_sol_xover4 lk_ball_wtd fa_elec hbond_sr_bb hbond_lr_bb hbond_bb_sc hbond_sc dslf_fa13 omega fa_dun p_aa_pp yhh_planarity ref rama_prepro cart_bonded total\\nweights 1 0.55 1 0.005 1 1 1 1 1 1 1 1.25 0.4 0.7 0.6 0.625 1 0.45 0.5 NA\\npose -30.6415 4.11697 22.1559 0.02551 0.15573 -1.34724 -8.89233 -6.0979 0 0 0 0 0.43483 1.11609 -0.99936 0 8.23965 -1.13484 5.52462 -7.34381\\nALA:NtermProteinFull_1 -1.99951 0.38956 1.04404 0.00274 0 -0.07479 -0.69491 0 0 0 0 0 -0.0498 0 0 0 1.32468 0 0.08677 0.02878\\nALA_2 -2.20479 0.3782 1.60514 0.00136 0 -0.03405 -0.74212 0 0 0 0 0 -0.00908 0 -0.3195 0 1.32468 -0.32611 0.15219 -0.17408\\nALA_3 -3.71815 0.46327 2.15032 0.00163 0 -0.22129 -0.62995 0 0 0 0 0 -0.05243 0 -0.34862 0 1.32468 -0.65294 0.28481 -1.39866\\nALA_4 -3.91444 0.80314 3.15125 0.00113 0 -0.12946 -1.69038 0 0 0 0 0 0.72584 0 0.23826 0 1.32468 0.57424 1.7981 2.88236\\nPRO_5 -4.55207 1.03472 2.61877 0.00267 0.03565 -0.08339 -1.02412 0 0 0 0 0 -0.1081 0.11242 0.08984 0 -1.64321 0.94059 1.90819 -0.66805\\nALA_6 -3.73418 0.28926 2.67094 0.00147 0 -0.12215 -0.66524 0 0 0 0 0 0.01867 0 -0.21709 0 1.32468 -0.14328 0.37456 -0.20237\\nLYS_7 -4.19343 0.31586 3.61223 0.00751 0.12008 -0.31578 -1.31242 0 0 0 0 0 -0.01615 1.00367 -0.11155 0 -0.71458 -0.2776 0.11137 -1.7708\\nALA_8 -2.64057 0.17208 1.99369 0.00153 0 -0.03822 -0.65111 0 0 0 0 0 -0.06019 0 -0.36386 0 1.32468 -0.4021 0.17043 -0.49364\\nALA_9 -2.03407 0.17052 1.62961 0.00237 0 -0.20444 -1.10969 0 0 0 0 0 -0.01392 0 0.03316 0 1.32468 -0.57746 0.29114 -0.48811\\nALA:CtermProteinFull_10 -1.65024 0.10036 1.67991 0.0031 0 -0.12367 -0.37238 0 0 0 0 0 0 0 0 0 1.32468 -0.27017 0.34706 1.03866\\n#END_POSE_ENERGIES_TABLE /tmp/tmp1bqdqyrg.pdb\\n\\n\",\"pdb\");\n\tviewer_17269384987144978.setStyle({\"stick\": {\"colorscheme\": \"greenCarbon\"}});\n\tviewer_17269384987144978.zoomTo();\nviewer_17269384987144978.render();\n});\n</script>","text/html":["<div id=\"3dmolviewer_17269384987144978\"  style=\"position: relative; width: 800px; height: 600px;\">\n","        <p id=\"3dmolwarning_17269384987144978\" style=\"background-color:#ffcccc;color:black\">3Dmol.js failed to load for some reason.  Please check your browser console for error messages.<br></p>\n","        </div>\n","<script>\n","\n","var loadScriptAsync = function(uri){\n","  return new Promise((resolve, reject) => {\n","    //this is to ignore the existence of requirejs amd\n","    var savedexports, savedmodule;\n","    if (typeof exports !== 'undefined') savedexports = exports;\n","    else exports = {}\n","    if (typeof module !== 'undefined') savedmodule = module;\n","    else module = {}\n","\n","    var tag = document.createElement('script');\n","    tag.src = uri;\n","    tag.async = true;\n","    tag.onload = () => {\n","        exports = savedexports;\n","        module = savedmodule;\n","        resolve();\n","    };\n","  var firstScriptTag = document.getElementsByTagName('script')[0];\n","  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n","});\n","};\n","\n","if(typeof $3Dmolpromise === 'undefined') {\n","$3Dmolpromise = null;\n","  $3Dmolpromise = loadScriptAsync('https://cdnjs.cloudflare.com/ajax/libs/3Dmol/2.4.0/3Dmol-min.js');\n","}\n","\n","var viewer_17269384987144978 = null;\n","var warn = document.getElementById(\"3dmolwarning_17269384987144978\");\n","if(warn) {\n","    warn.parentNode.removeChild(warn);\n","}\n","$3Dmolpromise.then(function() {\n","viewer_17269384987144978 = $3Dmol.createViewer(document.getElementById(\"3dmolviewer_17269384987144978\"),{backgroundColor:\"white\"});\n","viewer_17269384987144978.zoomTo();\n","\tviewer_17269384987144978.addModel(\"HEADER                                            21-SEP-24   XXXX              \\nEXPDTA    THEORETICAL MODEL                                                     \\nREMARK 220                                                                      \\nREMARK 220 EXPERIMENTAL DETAILS                                                 \\nREMARK 220  EXPERIMENT TYPE                : THEORETICAL MODELLING              \\nREMARK 220  DATE OF DATA COLLECTION        : 21-SEP-24                          \\nREMARK 220                                                                      \\nREMARK 220 REMARK: MODEL GENERATED BY ROSETTA                                   \\nREMARK 220  VERSION 2024.19+release.a34b73c40f                                  \\nATOM      1  N   ALA A   1       1.225  -1.236   0.557  1.00  0.00           N  \\nATOM      2  CA  ALA A   1       2.525  -0.732   0.129  1.00  0.00           C  \\nATOM      3  C   ALA A   1       2.543   0.778   0.102  1.00  0.00           C  \\nATOM      4  O   ALA A   1       1.984   1.385  -0.805  1.00  0.00           O  \\nATOM      5  CB  ALA A   1       2.879  -1.244  -1.257  1.00  0.00           C  \\nATOM      6 1H   ALA A   1       1.301  -2.219   0.775  1.00  0.00           H  \\nATOM      7 2H   ALA A   1       0.922  -0.734   1.382  1.00  0.00           H  \\nATOM      8 3H   ALA A   1       0.549  -1.106  -0.180  1.00  0.00           H  \\nATOM      9  HA  ALA A   1       3.274  -1.073   0.845  1.00  0.00           H  \\nATOM     10 1HB  ALA A   1       3.861  -0.867  -1.543  1.00  0.00           H  \\nATOM     11 2HB  ALA A   1       2.895  -2.333  -1.251  1.00  0.00           H  \\nATOM     12 3HB  ALA A   1       2.135  -0.897  -1.972  1.00  0.00           H  \\nATOM     13  N   ALA A   2       3.255   1.396   1.049  1.00  0.00           N  \\nATOM     14  CA  ALA A   2       3.368   2.865   1.116  1.00  0.00           C  \\nATOM     15  C   ALA A   2       3.969   3.415  -0.177  1.00  0.00           C  \\nATOM     16  O   ALA A   2       3.699   4.549  -0.596  1.00  0.00           O  \\nATOM     17  CB  ALA A   2       4.223   3.279   2.297  1.00  0.00           C  \\nATOM     18  H   ALA A   2       3.718   0.850   1.766  1.00  0.00           H  \\nATOM     19  HA  ALA A   2       2.366   3.283   1.231  1.00  0.00           H  \\nATOM     20 1HB  ALA A   2       4.278   4.364   2.340  1.00  0.00           H  \\nATOM     21 2HB  ALA A   2       3.779   2.900   3.217  1.00  0.00           H  \\nATOM     22 3HB  ALA A   2       5.224   2.869   2.180  1.00  0.00           H  \\nATOM     23  N   ALA A   3       4.811   2.601  -0.796  1.00  0.00           N  \\nATOM     24  CA  ALA A   3       5.449   2.944  -2.043  1.00  0.00           C  \\nATOM     25  C   ALA A   3       4.427   3.209  -3.139  1.00  0.00           C  \\nATOM     26  O   ALA A   3       4.688   4.001  -4.041  1.00  0.00           O  \\nATOM     27  CB  ALA A   3       6.366   1.824  -2.485  1.00  0.00           C  \\nATOM     28  H   ALA A   3       4.999   1.701  -0.386  1.00  0.00           H  \\nATOM     29  HA  ALA A   3       6.034   3.847  -1.886  1.00  0.00           H  \\nATOM     30 1HB  ALA A   3       6.858   2.105  -3.415  1.00  0.00           H  \\nATOM     31 2HB  ALA A   3       7.115   1.642  -1.717  1.00  0.00           H  \\nATOM     32 3HB  ALA A   3       5.780   0.920  -2.643  1.00  0.00           H  \\nATOM     33  N   ALA A   4       3.289   2.514  -3.126  1.00  0.00           N  \\nATOM     34  CA  ALA A   4       2.352   2.715  -4.210  1.00  0.00           C  \\nATOM     35  C   ALA A   4       1.868   4.188  -4.230  1.00  0.00           C  \\nATOM     36  O   ALA A   4       1.968   4.807  -5.293  1.00  0.00           O  \\nATOM     37  CB  ALA A   4       1.251   1.655  -4.157  1.00  0.00           C  \\nATOM     38  H   ALA A   4       3.045   1.903  -2.352  1.00  0.00           H  \\nATOM     39  HA  ALA A   4       2.882   2.560  -5.151  1.00  0.00           H  \\nATOM     40 1HB  ALA A   4       0.587   1.783  -5.010  1.00  0.00           H  \\nATOM     41 2HB  ALA A   4       1.728   0.676  -4.217  1.00  0.00           H  \\nATOM     42 3HB  ALA A   4       0.683   1.671  -3.274  1.00  0.00           H  \\nATOM     43  N   PRO A   5       1.306   4.798  -3.147  1.00  0.00           N  \\nATOM     44  CA  PRO A   5       1.035   6.225  -3.068  1.00  0.00           C  \\nATOM     45  C   PRO A   5       2.239   7.091  -3.403  1.00  0.00           C  \\nATOM     46  O   PRO A   5       2.097   8.134  -4.045  1.00  0.00           O  \\nATOM     47  CB  PRO A   5       0.652   6.400  -1.603  1.00  0.00           C  \\nATOM     48  CG  PRO A   5       0.020   5.103  -1.232  1.00  0.00           C  \\nATOM     49  CD  PRO A   5       0.799   4.058  -1.977  1.00  0.00           C  \\nATOM     50  HA  PRO A   5       0.185   6.463  -3.720  1.00  0.00           H  \\nATOM     51 1HB  PRO A   5       1.554   6.613  -1.005  1.00  0.00           H  \\nATOM     52 2HB  PRO A   5      -0.023   7.260  -1.488  1.00  0.00           H  \\nATOM     53 1HG  PRO A   5       0.080   4.956  -0.144  1.00  0.00           H  \\nATOM     54 2HG  PRO A   5      -1.045   5.107  -1.495  1.00  0.00           H  \\nATOM     55 1HD  PRO A   5       1.575   3.674  -1.376  1.00  0.00           H  \\nATOM     56 2HD  PRO A   5       0.071   3.323  -2.242  1.00  0.00           H  \\nATOM     57  N   ALA A   6       3.445   6.652  -3.005  1.00  0.00           N  \\nATOM     58  CA  ALA A   6       4.624   7.455  -3.306  1.00  0.00           C  \\nATOM     59  C   ALA A   6       4.821   7.585  -4.807  1.00  0.00           C  \\nATOM     60  O   ALA A   6       5.154   8.656  -5.315  1.00  0.00           O  \\nATOM     61  CB  ALA A   6       5.860   6.841  -2.676  1.00  0.00           C  \\nATOM     62  H   ALA A   6       3.546   5.799  -2.449  1.00  0.00           H  \\nATOM     63  HA  ALA A   6       4.468   8.452  -2.891  1.00  0.00           H  \\nATOM     64 1HB  ALA A   6       6.722   7.468  -2.890  1.00  0.00           H  \\nATOM     65 2HB  ALA A   6       5.717   6.763  -1.599  1.00  0.00           H  \\nATOM     66 3HB  ALA A   6       6.023   5.864  -3.093  1.00  0.00           H  \\nATOM     67  N   LYS A   7       4.590   6.485  -5.518  1.00  0.00           N  \\nATOM     68  CA  LYS A   7       4.719   6.468  -6.963  1.00  0.00           C  \\nATOM     69  C   LYS A   7       3.554   7.199  -7.613  1.00  0.00           C  \\nATOM     70  O   LYS A   7       3.739   7.951  -8.567  1.00  0.00           O  \\nATOM     71  CB  LYS A   7       4.805   5.027  -7.453  1.00  0.00           C  \\nATOM     72  CG  LYS A   7       6.112   4.323  -7.071  1.00  0.00           C  \\nATOM     73  CD  LYS A   7       6.132   2.877  -7.525  1.00  0.00           C  \\nATOM     74  CE  LYS A   7       7.429   2.190  -7.118  1.00  0.00           C  \\nATOM     75  NZ  LYS A   7       7.448   0.760  -7.530  1.00  0.00           N  \\nATOM     76  H   LYS A   7       4.341   5.623  -5.030  1.00  0.00           H  \\nATOM     77  HA  LYS A   7       5.639   6.986  -7.234  1.00  0.00           H  \\nATOM     78 1HB  LYS A   7       3.982   4.451  -7.021  1.00  0.00           H  \\nATOM     79 2HB  LYS A   7       4.696   4.996  -8.536  1.00  0.00           H  \\nATOM     80 1HG  LYS A   7       6.950   4.850  -7.526  1.00  0.00           H  \\nATOM     81 2HG  LYS A   7       6.233   4.358  -5.991  1.00  0.00           H  \\nATOM     82 1HD  LYS A   7       5.294   2.351  -7.061  1.00  0.00           H  \\nATOM     83 2HD  LYS A   7       6.019   2.827  -8.606  1.00  0.00           H  \\nATOM     84 1HE  LYS A   7       8.266   2.705  -7.587  1.00  0.00           H  \\nATOM     85 2HE  LYS A   7       7.538   2.249  -6.035  1.00  0.00           H  \\nATOM     86 1HZ  LYS A   7       8.321   0.338  -7.244  1.00  0.00           H  \\nATOM     87 2HZ  LYS A   7       6.679   0.275  -7.089  1.00  0.00           H  \\nATOM     88 3HZ  LYS A   7       7.357   0.696  -8.533  1.00  0.00           H  \\nATOM     89  N   ALA A   8       2.354   7.020  -7.065  1.00  0.00           N  \\nATOM     90  CA  ALA A   8       1.147   7.646  -7.599  1.00  0.00           C  \\nATOM     91  C   ALA A   8       1.260   9.164  -7.597  1.00  0.00           C  \\nATOM     92  O   ALA A   8       0.762   9.847  -8.492  1.00  0.00           O  \\nATOM     93  CB  ALA A   8      -0.059   7.231  -6.780  1.00  0.00           C  \\nATOM     94  H   ALA A   8       2.262   6.387  -6.270  1.00  0.00           H  \\nATOM     95  HA  ALA A   8       1.022   7.312  -8.629  1.00  0.00           H  \\nATOM     96 1HB  ALA A   8      -0.955   7.680  -7.201  1.00  0.00           H  \\nATOM     97 2HB  ALA A   8      -0.149   6.145  -6.792  1.00  0.00           H  \\nATOM     98 3HB  ALA A   8       0.075   7.574  -5.763  1.00  0.00           H  \\nATOM     99  N   ALA A   9       1.934   9.685  -6.577  1.00  0.00           N  \\nATOM    100  CA  ALA A   9       2.145  11.106  -6.364  1.00  0.00           C  \\nATOM    101  C   ALA A   9       3.315  11.694  -7.167  1.00  0.00           C  \\nATOM    102  O   ALA A   9       3.591  12.890  -7.037  1.00  0.00           O  \\nATOM    103  CB  ALA A   9       2.399  11.343  -4.889  1.00  0.00           C  \\nATOM    104  H   ALA A   9       2.288   9.050  -5.861  1.00  0.00           H  \\nATOM    105  HA  ALA A   9       1.237  11.626  -6.666  1.00  0.00           H  \\nATOM    106 1HB  ALA A   9       2.520  12.407  -4.702  1.00  0.00           H  \\nATOM    107 2HB  ALA A   9       1.558  10.961  -4.309  1.00  0.00           H  \\nATOM    108 3HB  ALA A   9       3.309  10.811  -4.598  1.00  0.00           H  \\nATOM    109  N   ALA A  10       4.025  10.864  -7.937  1.00  0.00           N  \\nATOM    110  CA  ALA A  10       5.200  11.305  -8.675  1.00  0.00           C  \\nATOM    111  C   ALA A  10       4.915  11.470 -10.169  1.00  0.00           C  \\nATOM    112  O   ALA A  10       4.627  12.579 -10.624  1.00  0.00           O  \\nATOM    113  OXT ALA A  10       5.351  10.611 -10.936  1.00  0.00           O  \\nATOM    114  CB  ALA A  10       6.324  10.302  -8.462  1.00  0.00           C  \\nATOM    115  H   ALA A  10       3.750   9.886  -8.025  1.00  0.00           H  \\nATOM    116  HA  ALA A  10       5.507  12.273  -8.282  1.00  0.00           H  \\nATOM    117 1HB  ALA A  10       7.220  10.636  -8.982  1.00  0.00           H  \\nATOM    118 2HB  ALA A  10       6.532  10.213  -7.394  1.00  0.00           H  \\nATOM    119 3HB  ALA A  10       6.017   9.330  -8.851  1.00  0.00           H  \\nTER                                                                             \\n# All scores below are weighted scores, not raw scores.\\n#BEGIN_POSE_ENERGIES_TABLE /tmp/tmp1bqdqyrg.pdb\\nlabel fa_atr fa_rep fa_sol fa_intra_rep fa_intra_sol_xover4 lk_ball_wtd fa_elec hbond_sr_bb hbond_lr_bb hbond_bb_sc hbond_sc dslf_fa13 omega fa_dun p_aa_pp yhh_planarity ref rama_prepro cart_bonded total\\nweights 1 0.55 1 0.005 1 1 1 1 1 1 1 1.25 0.4 0.7 0.6 0.625 1 0.45 0.5 NA\\npose -30.6415 4.11697 22.1559 0.02551 0.15573 -1.34724 -8.89233 -6.0979 0 0 0 0 0.43483 1.11609 -0.99936 0 8.23965 -1.13484 5.52462 -7.34381\\nALA:NtermProteinFull_1 -1.99951 0.38956 1.04404 0.00274 0 -0.07479 -0.69491 0 0 0 0 0 -0.0498 0 0 0 1.32468 0 0.08677 0.02878\\nALA_2 -2.20479 0.3782 1.60514 0.00136 0 -0.03405 -0.74212 0 0 0 0 0 -0.00908 0 -0.3195 0 1.32468 -0.32611 0.15219 -0.17408\\nALA_3 -3.71815 0.46327 2.15032 0.00163 0 -0.22129 -0.62995 0 0 0 0 0 -0.05243 0 -0.34862 0 1.32468 -0.65294 0.28481 -1.39866\\nALA_4 -3.91444 0.80314 3.15125 0.00113 0 -0.12946 -1.69038 0 0 0 0 0 0.72584 0 0.23826 0 1.32468 0.57424 1.7981 2.88236\\nPRO_5 -4.55207 1.03472 2.61877 0.00267 0.03565 -0.08339 -1.02412 0 0 0 0 0 -0.1081 0.11242 0.08984 0 -1.64321 0.94059 1.90819 -0.66805\\nALA_6 -3.73418 0.28926 2.67094 0.00147 0 -0.12215 -0.66524 0 0 0 0 0 0.01867 0 -0.21709 0 1.32468 -0.14328 0.37456 -0.20237\\nLYS_7 -4.19343 0.31586 3.61223 0.00751 0.12008 -0.31578 -1.31242 0 0 0 0 0 -0.01615 1.00367 -0.11155 0 -0.71458 -0.2776 0.11137 -1.7708\\nALA_8 -2.64057 0.17208 1.99369 0.00153 0 -0.03822 -0.65111 0 0 0 0 0 -0.06019 0 -0.36386 0 1.32468 -0.4021 0.17043 -0.49364\\nALA_9 -2.03407 0.17052 1.62961 0.00237 0 -0.20444 -1.10969 0 0 0 0 0 -0.01392 0 0.03316 0 1.32468 -0.57746 0.29114 -0.48811\\nALA:CtermProteinFull_10 -1.65024 0.10036 1.67991 0.0031 0 -0.12367 -0.37238 0 0 0 0 0 0 0 0 0 1.32468 -0.27017 0.34706 1.03866\\n#END_POSE_ENERGIES_TABLE /tmp/tmp1bqdqyrg.pdb\\n\\n\",\"pdb\");\n","\tviewer_17269384987144978.setStyle({\"stick\": {\"colorscheme\": \"greenCarbon\"}});\n","\tviewer_17269384987144978.zoomTo();\n","viewer_17269384987144978.render();\n","});\n","</script>"]},"metadata":{}}]},{"cell_type":"markdown","id":"733365f6","metadata":{"id":"733365f6"},"source":["Since the peptide was already folded, we don't see major changes in its conformation. However, we can compare the relative stability of the design by measuring the Rosetta energy before and after design.\n"]},{"cell_type":"code","execution_count":null,"id":"0e12686c","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"0e12686c","outputId":"55149293-48e1-4d6c-c7b6-ae277600d983"},"outputs":[{"output_type":"stream","name":"stdout","text":["core.energy_methods.CartesianBondedEnergy: Creating new peptide-bonded energy container (10)\n","-5.371127246023434\n","-7.343814043773602\n"]}],"source":["before_design = scorefxn(pose)\n","print(before_design)\n","energy_design = scorefxn(d3l7_pose)\n","print(energy_design)"]},{"cell_type":"markdown","source":["**Exercise:** Apparently, we have stabilised the helix with our design. How is this possible?"],"metadata":{"id":"OcaMNAj9kBJR"},"id":"OcaMNAj9kBJR"},{"cell_type":"markdown","id":"4435f7c5","metadata":{"id":"4435f7c5"},"source":["To properly measure the energy level of the no-design peptide, we need to subject it to the same energy minimization.\n"]},{"cell_type":"code","execution_count":null,"id":"1ba0e1cb","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"1ba0e1cb","outputId":"fb7d608b-a335-4e9b-92a8-3d6d11afa768"},"outputs":[{"output_type":"stream","name":"stdout","text":["core.pack.task: Packer task: initialize from command line()\n","#Packer_Task\n","\n","Threads to request: ALL AVAILABLE\n","\n","resid\tpack?\tdesign?\tallowed_aas\n","1\tTRUE\tFALSE\tALA:NtermProteinFull\n","2\tTRUE\tFALSE\tALA\n","3\tTRUE\tFALSE\tALA\n","4\tTRUE\tFALSE\tALA\n","5\tTRUE\tFALSE\tALA\n","6\tTRUE\tFALSE\tALA\n","7\tTRUE\tFALSE\tALA\n","8\tTRUE\tFALSE\tALA\n","9\tTRUE\tFALSE\tALA\n","10\tTRUE\tFALSE\tALA:CtermProteinFull\n","\n","core.energy_methods.CartesianBondedEnergy: Creating new peptide-bonded energy container (10)\n","protocols.relax.FastRelax: CMD: repeat  -5.37113  0  0  0.55\n","protocols.relax.FastRelax: CMD: coord_cst_weight  -5.37113  0  0  0.55\n","protocols.relax.FastRelax: CMD: scale:fa_rep  -7.87213  0  0  0.03245\n","core.pack.task: Packer task: initialize from command line()\n","core.pack.pack_rotamers: built 10 rotamers at 10 positions.\n","core.pack.interaction_graph.interaction_graph_factory: Instantiating DensePDInteractionGraph\n","protocols.relax.FastRelax: CMD: repack  -7.87213  0  0  0.03245\n","protocols.relax.FastRelax: CMD: scale:fa_rep  -7.78442  0  0  0.0506\n","protocols.relax.FastRelax: CMD: min  -14.8919  0.418776  0.418776  0.0506\n","protocols.relax.FastRelax: CMD: coord_cst_weight  -14.8919  0.418776  0.418776  0.0506\n","protocols.relax.FastRelax: CMD: scale:fa_rep  -13.6261  0.418776  0.418776  0.154\n","core.pack.task: Packer task: initialize from command line()\n","core.pack.pack_rotamers: built 10 rotamers at 10 positions.\n","core.pack.interaction_graph.interaction_graph_factory: Instantiating DensePDInteractionGraph\n","protocols.relax.FastRelax: CMD: repack  -13.6261  0.418776  0.418776  0.154\n","protocols.relax.FastRelax: CMD: scale:fa_rep  -13.3366  0.418776  0.418776  0.17765\n","protocols.relax.FastRelax: CMD: min  -13.4572  0.41783  0.41783  0.17765\n","protocols.relax.FastRelax: CMD: coord_cst_weight  -13.4572  0.41783  0.41783  0.17765\n","protocols.relax.FastRelax: CMD: scale:fa_rep  -12.0415  0.41783  0.41783  0.3124\n","core.pack.task: Packer task: initialize from command line()\n","core.pack.pack_rotamers: built 10 rotamers at 10 positions.\n","core.pack.interaction_graph.interaction_graph_factory: Instantiating DensePDInteractionGraph\n","protocols.relax.FastRelax: CMD: repack  -12.0415  0.41783  0.41783  0.3124\n","protocols.relax.FastRelax: CMD: scale:fa_rep  -11.666  0.41783  0.41783  0.34815\n","protocols.relax.FastRelax: CMD: min  -12.6772  0.324238  0.324238  0.34815\n","protocols.relax.FastRelax: CMD: coord_cst_weight  -12.6772  0.324238  0.324238  0.34815\n","protocols.relax.FastRelax: CMD: scale:fa_rep  -11.5137  0.324238  0.324238  0.55\n","core.pack.task: Packer task: initialize from command line()\n","core.pack.pack_rotamers: built 10 rotamers at 10 positions.\n","core.pack.interaction_graph.interaction_graph_factory: Instantiating DensePDInteractionGraph\n","protocols.relax.FastRelax: CMD: repack  -11.5137  0.324238  0.324238  0.55\n","protocols.relax.FastRelax: CMD: min  -11.6882  0.297111  0.297111  0.55\n","protocols.relax.FastRelax: MRP: 0  -11.6882  -11.6882  0.297111  0.297111\n","protocols.relax.FastRelax: CMD: accept_to_best  -11.6882  0.297111  0.297111  0.55\n","protocols.relax.FastRelax: CMD: endrepeat  -11.6882  0.297111  0.297111  0.55\n","protocols::checkpoint: Deleting checkpoints of FastRelax\n","CPU times: user 711 ms, sys: 9.7 ms, total: 720 ms\n","Wall time: 747 ms\n"]}],"source":["# new design resfile\n","nodesign_resfile = \"/content/google_drive/MyDrive/BIP_24-25/5-Protein_design/nodesign.resfile\"\n","\n","#reset pose to the initial conformation\n","pose = start_pose.clone()\n","\n","\n","# The task factory accepts all the task operations\n","tf = pyrosetta.rosetta.core.pack.task.TaskFactory()\n","\n","# These are pretty standard\n","tf.push_back(pyrosetta.rosetta.core.pack.task.operation.InitializeFromCommandline())\n","tf.push_back(pyrosetta.rosetta.core.pack.task.operation.IncludeCurrent())\n","\n","\n","# Include the resfile\n","tf.push_back(pyrosetta.rosetta.core.pack.task.operation.ReadResfile(nodesign_resfile))\n","\n","# Convert the task factory into a PackerTask to take a look at it\n","packer_task = tf.create_task_and_apply_taskoperations(pose)\n","# View the PackerTask\n","print(packer_task)\n","\n","\n","mm = pyrosetta.rosetta.core.kinematics.MoveMap()\n","mm.set_bb(True)\n","mm.set_chi(True)\n","mm.set_jump(True)\n","\n","rel_design = pyrosetta.rosetta.protocols.relax.FastRelax(scorefxn_in=scorefxn, standard_repeats=1, script_file=\"MonomerDesign2019\")\n","rel_design.cartesian(True)\n","rel_design.set_task_factory(tf)\n","rel_design.set_movemap(mm)\n","rel_design.minimize_bond_angles(True)\n","rel_design.minimize_bond_lengths(True)\n","\n","%time rel_design.apply(pose)\n","\n"]},{"cell_type":"code","execution_count":null,"id":"f9746fef","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"f9746fef","outputId":"febabe5e-4271-4151-c3be-28c5083490f3"},"outputs":[{"output_type":"stream","name":"stdout","text":["-11.688240255929747\n"]}],"source":["energy_nodesign = scorefxn(pose)\n","print(energy_nodesign)"]},{"cell_type":"code","execution_count":null,"id":"b10e9aee","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"b10e9aee","outputId":"6429b131-b237-4cbb-e252-25836553454f"},"outputs":[{"output_type":"stream","name":"stdout","text":["deltaE = 4.344426212156145\n"]}],"source":["print(\"deltaE =\", energy_design - energy_nodesign)"]},{"cell_type":"markdown","id":"30b31db5","metadata":{"id":"30b31db5"},"source":["**Exercise. Mess with the polyA peptide. Explore the mutations that destabilize the $\\alpha$-helix.**"]}],"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.11.5"},"colab":{"provenance":[]}},"nbformat":4,"nbformat_minor":5}